<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Shane Hogle">
<meta name="dcterms.date" content="2024-12-13">

<title>omicOceansEnvData - Harmonize environmental data and sequencing identifiers</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<link href="../../site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="../../site_libs/pagedtable-1.1/js/pagedtable.js"></script>


</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../R/tara/01_download_data_ncbi.html">1. Tara oceans workflow</a></li><li class="breadcrumb-item"><a href="../../R/tara/03_harmonize_ncbi_pangaea.html">3) Harmonize env and ’omics data</a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../../">omicOceansEnvData</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">1. Tara oceans workflow</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../R/tara/01_download_data_ncbi.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">1) BioSamples at NCBI</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../R/tara/02_download_data_pangaea.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">2) Env data at Pangaea.de</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../R/tara/03_harmonize_ncbi_pangaea.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">3) Harmonize env and ’omics data</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#setup" id="toc-setup" class="nav-link active" data-scroll-target="#setup"><span class="header-section-number">1</span> Setup</a></li>
  <li><a href="#read-data" id="toc-read-data" class="nav-link" data-scroll-target="#read-data"><span class="header-section-number">2</span> Read data</a></li>
  <li><a href="#collate-filter-and-harmonize-environmental-data" id="toc-collate-filter-and-harmonize-environmental-data" class="nav-link" data-scroll-target="#collate-filter-and-harmonize-environmental-data"><span class="header-section-number">3</span> Collate, filter, and harmonize environmental data</a>
  <ul class="collapse">
  <li><a href="#combine-all-pangaea-datasets-with-biosamples" id="toc-combine-all-pangaea-datasets-with-biosamples" class="nav-link" data-scroll-target="#combine-all-pangaea-datasets-with-biosamples"><span class="header-section-number">3.1</span> Combine all Pangaea datasets with BioSamples</a></li>
  <li><a href="#fill-missing-observations-with-adjacent-data" id="toc-fill-missing-observations-with-adjacent-data" class="nav-link" data-scroll-target="#fill-missing-observations-with-adjacent-data"><span class="header-section-number">3.2</span> Fill missing observations with adjacent data</a></li>
  <li><a href="#fill-missing-observations-with-whole-water-column-data-pangaea.875579" id="toc-fill-missing-observations-with-whole-water-column-data-pangaea.875579" class="nav-link" data-scroll-target="#fill-missing-observations-with-whole-water-column-data-pangaea.875579"><span class="header-section-number">3.3</span> Fill missing observations with “whole water column” data (PANGAEA.875579)</a></li>
  <li><a href="#fill-missing-observations-with-darwin-model-nutrients-output" id="toc-fill-missing-observations-with-darwin-model-nutrients-output" class="nav-link" data-scroll-target="#fill-missing-observations-with-darwin-model-nutrients-output"><span class="header-section-number">3.4</span> Fill missing observations with Darwin model Nutrients output</a>
  <ul class="collapse">
  <li><a href="#darwin-coordinates-for-nutrients" id="toc-darwin-coordinates-for-nutrients" class="nav-link" data-scroll-target="#darwin-coordinates-for-nutrients"><span class="header-section-number">3.4.1</span> Darwin coordinates for nutrients</a></li>
  <li><a href="#filter-darwin-output-to-tara-coordinates" id="toc-filter-darwin-output-to-tara-coordinates" class="nav-link" data-scroll-target="#filter-darwin-output-to-tara-coordinates"><span class="header-section-number">3.4.2</span> Filter Darwin output to Tara coordinates</a></li>
  <li><a href="#join-observed-and-darwin-model-output-into-single-dataframe" id="toc-join-observed-and-darwin-model-output-into-single-dataframe" class="nav-link" data-scroll-target="#join-observed-and-darwin-model-output-into-single-dataframe"><span class="header-section-number">3.4.3</span> Join observed and Darwin Model output into single dataframe</a></li>
  </ul></li>
  <li><a href="#impute-remaining-missing-observations" id="toc-impute-remaining-missing-observations" class="nav-link" data-scroll-target="#impute-remaining-missing-observations"><span class="header-section-number">3.5</span> Impute remaining missing observations</a></li>
  <li><a href="#joining-and-replacing-missing-values-with-those-imputed" id="toc-joining-and-replacing-missing-values-with-those-imputed" class="nav-link" data-scroll-target="#joining-and-replacing-missing-values-with-those-imputed"><span class="header-section-number">3.6</span> Joining and replacing missing values with those imputed</a>
  <ul class="collapse">
  <li><a href="#detour-to-add-some-additional-geographic-info" id="toc-detour-to-add-some-additional-geographic-info" class="nav-link" data-scroll-target="#detour-to-add-some-additional-geographic-info"><span class="header-section-number">3.6.1</span> Detour to add some additional geographic info</a></li>
  <li><a href="#finalize" id="toc-finalize" class="nav-link" data-scroll-target="#finalize"><span class="header-section-number">3.6.2</span> Finalize</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../R/tara/01_download_data_ncbi.html">1. Tara oceans workflow</a></li><li class="breadcrumb-item"><a href="../../R/tara/03_harmonize_ncbi_pangaea.html">3) Harmonize env and ’omics data</a></li></ol></nav>
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Harmonize environmental data and sequencing identifiers</h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
<p class="subtitle lead">Tara Oceans workflow</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Shane Hogle </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">December 13, 2024</p>
    </div>
  </div>
  
    
  </div>
  
<div>
  <div class="abstract">
    <div class="block-title">Abstract</div>
    Code in this notebook collates environmental parameters from different Pangaea.de datasets with the NCBI sequencing datasets. It then fills missing values from Pangaea with Darwin Model output for the nearest coordinate/depth/date. Finally, a small number of observations are imputed. The goal is to be able to associate an NCBI sequencing run accession number with environmental data for later statistical analysis. We do not want any missing data so we must try and patch missing data from other sources.
  </div>
</div>


</header>


<section id="setup" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Setup</h1>
<p>This code loads required libraries and sets global variables</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="fu">library</span>(here)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="fu">library</span>(lubridate)</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="fu">library</span>(fuzzyjoin)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="read-data" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Read data</h1>
<p>Read NCBI BioSample data with unique Tara Ocean Barcodes</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>tara_biosamples <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_tsv</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data"</span>, <span class="st">"tara"</span>, <span class="st">"ncbi2tara_barcode.tsv"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Read previously saved environmental data and metadata downloaded from Pangaea.de</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co"># tara stations</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>PANGAEA_842237 <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_rds</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"_data_raw"</span>, <span class="st">"tara"</span>, <span class="st">"PANGAEA_842237.rds"</span>))</span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="co"># carbonate chemistry</span></span>
<span id="cb3-4"><a href="#cb3-4"></a>PANGAEA_875567 <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_rds</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"_data_raw"</span>, <span class="st">"tara"</span>, <span class="st">"PANGAEA_875567.rds"</span>))</span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="co"># nutrients</span></span>
<span id="cb3-6"><a href="#cb3-6"></a>PANGAEA_875575 <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_rds</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"_data_raw"</span>, <span class="st">"tara"</span>, <span class="st">"PANGAEA_875575.rds"</span>))</span>
<span id="cb3-7"><a href="#cb3-7"></a><span class="co"># phytoplankton pigments</span></span>
<span id="cb3-8"><a href="#cb3-8"></a>PANGAEA_875569 <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_rds</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"_data_raw"</span>, <span class="st">"tara"</span>, <span class="st">"PANGAEA_875569.rds"</span>))</span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="co"># CTD sensor data</span></span>
<span id="cb3-10"><a href="#cb3-10"></a>PANGAEA_875576 <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_rds</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"_data_raw"</span>, <span class="st">"tara"</span>, <span class="st">"PANGAEA_875576.rds"</span>))</span>
<span id="cb3-11"><a href="#cb3-11"></a><span class="co"># mesoscale features</span></span>
<span id="cb3-12"><a href="#cb3-12"></a>PANGAEA_875577 <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_rds</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"_data_raw"</span>, <span class="st">"tara"</span>, <span class="st">"PANGAEA_875577.rds"</span>))</span>
<span id="cb3-13"><a href="#cb3-13"></a><span class="co"># whole water column features</span></span>
<span id="cb3-14"><a href="#cb3-14"></a>PANGAEA_875579 <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_rds</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"_data_raw"</span>, <span class="st">"tara"</span>, <span class="st">"PANGAEA_875579.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="collate-filter-and-harmonize-environmental-data" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Collate, filter, and harmonize environmental data</h1>
<p>In the final metagenomic tables we are interested in the following environmental variables:</p>
<ol type="1">
<li>Density</li>
<li>Temperature</li>
<li>Salinity</li>
<li>Chlorophyll a</li>
<li>Dissolved oxygen (O2)</li>
<li>Total inorganic nitrogen (NO3 + NO2)</li>
<li>Nitrite (NO2)</li>
<li>Silica (SiO2)</li>
<li>Phosphate (PO4)</li>
<li>Total dissolved iron (FeT)</li>
</ol>
<p>Thus we need to access the following environmental variables from various Pangaea dataframes for the final product</p>
<table class="table">
<colgroup>
<col style="width: 46%">
<col style="width: 34%">
<col style="width: 19%">
</colgroup>
<thead>
<tr class="header">
<th>Variable</th>
<th>Dataset</th>
<th>Variable</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Density</td>
<td>Sensors, in situ (PANGAEA.875576)</td>
<td>PANGAEA.875576_047</td>
</tr>
<tr class="even">
<td>Salinity</td>
<td>Sensors, in situ (PANGAEA.875576)</td>
<td>PANGAEA.875576_037</td>
</tr>
<tr class="odd">
<td>Temperature</td>
<td>Sensors, in situ (PANGAEA.875576)</td>
<td>PANGAEA.875576_027</td>
</tr>
<tr class="even">
<td>Dissolved Oxygen</td>
<td>Sensors, in situ (PANGAEA.875576)</td>
<td>PANGAEA.875576_057</td>
</tr>
<tr class="odd">
<td>Chlorophyll a</td>
<td>Sensors, in situ (PANGAEA.875576)</td>
<td>PANGAEA.875576_077</td>
</tr>
<tr class="even">
<td>Dissolved inorganic nitrogen (DIN = NO3 + NO2)</td>
<td>Nutrients, bottles (PANGAEA.875575)</td>
<td>PANGAEA.875575_031</td>
</tr>
<tr class="odd">
<td>Nitrite (NO2)</td>
<td>Nutrients, bottles (PANGAEA.875575)</td>
<td>PANGAEA.875575_021</td>
</tr>
<tr class="even">
<td>Phosphate (PO4)</td>
<td>Nutrients, bottles (PANGAEA.875575)</td>
<td>PANGAEA.875575_026</td>
</tr>
<tr class="odd">
<td>Silica (SiO2)</td>
<td>Nutrients, bottles (PANGAEA.875575)</td>
<td>PANGAEA.875575_036</td>
</tr>
<tr class="even">
<td>Darwin Model Total dissolved Iron (FeT)</td>
<td>Modeled variables (PANGAEA.875577)</td>
<td>PANGAEA.875577_039</td>
</tr>
<tr class="odd">
<td>Darwin Model NO2</td>
<td>Modeled variables (PANGAEA.875577)</td>
<td>PANGAEA.875577_043</td>
</tr>
<tr class="even">
<td>Darwin Model DIN</td>
<td>Modeled variables (PANGAEA.875577)</td>
<td>PANGAEA.875577_045</td>
</tr>
<tr class="odd">
<td>Temp @ 10 meters</td>
<td>Water column (PANGAEA.875579)</td>
<td>PANGAEA.875579_054</td>
</tr>
<tr class="even">
<td>Temp @ mixed layer</td>
<td>Water column (PANGAEA.875579)</td>
<td>PANGAEA.875579_057</td>
</tr>
<tr class="odd">
<td>Temp @ DCM</td>
<td>Water column (PANGAEA.875579)</td>
<td>PANGAEA.875579_059</td>
</tr>
<tr class="even">
<td>Salinity @ 10 meters</td>
<td>Water column (PANGAEA.875579)</td>
<td>PANGAEA.875579_064</td>
</tr>
<tr class="odd">
<td>Salinity @ mixed layer</td>
<td>Water column (PANGAEA.875579)</td>
<td>PANGAEA.875579_067</td>
</tr>
<tr class="even">
<td>Salinity @ DCM</td>
<td>Water column (PANGAEA.875579)</td>
<td>PANGAEA.875579_069</td>
</tr>
<tr class="odd">
<td>Density @ 10 meters</td>
<td>Water column (PANGAEA.875579)</td>
<td>PANGAEA.875579_074</td>
</tr>
<tr class="even">
<td>Density @ mixed layer</td>
<td>Water column (PANGAEA.875579)</td>
<td>PANGAEA.875579_077</td>
</tr>
<tr class="odd">
<td>Density @ DCM</td>
<td>Water column (PANGAEA.875579)</td>
<td>PANGAEA.875579_079</td>
</tr>
<tr class="even">
<td>Chlorophyll a @ 10 meters</td>
<td>Water column (PANGAEA.875579)</td>
<td>PANGAEA.875579_094</td>
</tr>
<tr class="odd">
<td>Chlorophyll a @ mixed layer</td>
<td>Water column (PANGAEA.875579)</td>
<td>PANGAEA.875579_097</td>
</tr>
<tr class="even">
<td>Chlorophyll a @ DCM</td>
<td>Water column (PANGAEA.875579)</td>
<td>PANGAEA.875579_099</td>
</tr>
<tr class="odd">
<td>Dissolved Oxygen @ 10 meters</td>
<td>Water column (PANGAEA.875579)</td>
<td>PANGAEA.875579_104</td>
</tr>
<tr class="even">
<td>Dissolved Oxygen @ mixed layer</td>
<td>Water column (PANGAEA.875579)</td>
<td>PANGAEA.875579_107</td>
</tr>
<tr class="odd">
<td>Dissolved Oxygen @ DCM</td>
<td>Water column (PANGAEA.875579)</td>
<td>PANGAEA.875579_109</td>
</tr>
<tr class="even">
<td>DIN @ 10 meters</td>
<td>Water column (PANGAEA.875579)</td>
<td>PANGAEA.875579_114</td>
</tr>
<tr class="odd">
<td>DIN @ mixed layer</td>
<td>Water column (PANGAEA.875579)</td>
<td>PANGAEA.875579_117</td>
</tr>
<tr class="even">
<td>DIN @ DCM</td>
<td>Water column (PANGAEA.875579)</td>
<td>PANGAEA.875579_119</td>
</tr>
</tbody>
</table>
<section id="combine-all-pangaea-datasets-with-biosamples" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="combine-all-pangaea-datasets-with-biosamples"><span class="header-section-number">3.1</span> Combine all Pangaea datasets with BioSamples</h2>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>tara_biosamples_joined01 <span class="ot">&lt;-</span> <span class="fu">left_join</span>(tara_biosamples, PANGAEA_875575, <span class="at">by =</span> <span class="fu">join_by</span>(tara_barcode_num, tara_station)) <span class="sc">%&gt;%</span> </span>
<span id="cb4-2"><a href="#cb4-2"></a>  <span class="fu">left_join</span>(dplyr<span class="sc">::</span><span class="fu">select</span>(PANGAEA_875576, tara_barcode_num, tara_station, date_time, <span class="dv">21</span><span class="sc">:</span><span class="fu">last_col</span>()), </span>
<span id="cb4-3"><a href="#cb4-3"></a>            <span class="at">by =</span> <span class="fu">join_by</span>(tara_barcode_num, tara_station)) <span class="sc">%&gt;%</span> </span>
<span id="cb4-4"><a href="#cb4-4"></a>  <span class="fu">left_join</span>(dplyr<span class="sc">::</span><span class="fu">select</span>(PANGAEA_875577, tara_barcode_num, tara_station, <span class="dv">21</span><span class="sc">:</span><span class="fu">last_col</span>()), <span class="at">by =</span> <span class="fu">join_by</span>(tara_barcode_num, tara_station)) <span class="sc">%&gt;%</span> </span>
<span id="cb4-5"><a href="#cb4-5"></a>  <span class="fu">left_join</span>(dplyr<span class="sc">::</span><span class="fu">select</span>(PANGAEA_875579, tara_barcode_num, tara_station, <span class="dv">21</span><span class="sc">:</span><span class="fu">last_col</span>()), <span class="at">by =</span> <span class="fu">join_by</span>(tara_barcode_num, tara_station)) <span class="sc">%&gt;%</span> </span>
<span id="cb4-6"><a href="#cb4-6"></a>  <span class="fu">left_join</span>(dplyr<span class="sc">::</span><span class="fu">select</span>(PANGAEA_842237, tara_station, latitude, longitude), <span class="at">by =</span> <span class="fu">join_by</span>(tara_station))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="fill-missing-observations-with-adjacent-data" class="level2" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="fill-missing-observations-with-adjacent-data"><span class="header-section-number">3.2</span> Fill missing observations with adjacent data</h2>
<p>There are 7 cases where there is not single recorded depth for an observation, but rather a depth range (<code>depth_min</code> to <code>depth_max</code>).</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>tara_biosamples_joined01 <span class="sc">%&gt;%</span> </span>
<span id="cb5-2"><a href="#cb5-2"></a>  <span class="fu">filter</span>(<span class="fu">is.na</span>(depth)) <span class="sc">%&gt;%</span> </span>
<span id="cb5-3"><a href="#cb5-3"></a>  <span class="fu">distinct</span>(tara_barcode_num, tara_station, env_ontology, depth, depth_min, depth_max, date_time)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["tara_barcode_num"],"name":[1],"type":["chr"],"align":["left"]},{"label":["tara_station"],"name":[2],"type":["chr"],"align":["left"]},{"label":["env_ontology"],"name":[3],"type":["chr"],"align":["left"]},{"label":["depth"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["depth_min"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["depth_max"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["date_time"],"name":[7],"type":["dttm"],"align":["right"]}],"data":[{"1":"TARA_A100000810","2":"TARA_019","3":"[ZZZ] marine water layer (ENVO:01000295)","4":"NA","5":"0","6":"100","7":"2009-11-11 18:27:00"},{"1":"TARA_A100000693","2":"TARA_021","3":"[ZZZ] marine water layer (ENVO:01000295)","4":"NA","5":"0","6":"500","7":"2009-11-15 10:31:00"},{"1":"TARA_B100002029","2":"TARA_152","3":"[MIX] marine epipelagic wind mixed layer (ENVO:01000061)","4":"NA","5":"5","6":"160","7":"2012-03-19 18:46:00"},{"1":"TARA_B110000037","2":"TARA_168","3":"[MIX] marine epipelagic wind mixed layer (ENVO:01000061)","4":"NA","5":"10","6":"100","7":"2013-07-02 04:24:00"},{"1":"TARA_R110002074","2":"TARA_168","3":"[MIX] marine epipelagic wind mixed layer (ENVO:01000061)","4":"NA","5":"10","6":"100","7":"2013-07-02 04:24:00"},{"1":"TARA_B110000093","2":"TARA_175","3":"[MIX] marine epipelagic wind mixed layer (ENVO:01000061)","4":"NA","5":"10","6":"100","7":"2013-07-11 06:13:00"},{"1":"TARA_R110002094","2":"TARA_175","3":"[MIX] marine epipelagic wind mixed layer (ENVO:01000061)","4":"NA","5":"10","6":"100","7":"2013-07-11 06:13:00"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Later it will make our lives easier if there are actual values for these depths and not NAs. We fill these depths with the mean of the depth range.</p>
<p>Also in some cases nutrients or sensor data will not be joined even though there is a measurement at that particular station/depth. We will group by station and depth and then fill those observations from the same station/depth.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>tara_biosamples_joined02 <span class="ot">&lt;-</span> tara_biosamples_joined01 <span class="sc">%&gt;%</span> </span>
<span id="cb6-2"><a href="#cb6-2"></a>  <span class="fu">mutate</span>(<span class="at">depth =</span> <span class="fu">if_else</span>(<span class="fu">is.na</span>(depth), (depth_max <span class="sc">-</span> depth_min)<span class="sc">/</span><span class="dv">2</span>, depth)) <span class="sc">%&gt;%</span> </span>
<span id="cb6-3"><a href="#cb6-3"></a>  <span class="fu">arrange</span>(tara_station, depth) <span class="sc">%&gt;%</span> </span>
<span id="cb6-4"><a href="#cb6-4"></a>  <span class="fu">group_by</span>(tara_station, depth, depth_min, depth_max) <span class="sc">%&gt;%</span> </span>
<span id="cb6-5"><a href="#cb6-5"></a>  <span class="fu">fill</span>(<span class="dv">20</span><span class="sc">:</span><span class="fu">last_col</span>(), <span class="at">.direction =</span> <span class="st">"downup"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb6-6"><a href="#cb6-6"></a>  <span class="fu">ungroup</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="fill-missing-observations-with-whole-water-column-data-pangaea.875579" class="level2" data-number="3.3">
<h2 data-number="3.3" class="anchored" data-anchor-id="fill-missing-observations-with-whole-water-column-data-pangaea.875579"><span class="header-section-number">3.3</span> Fill missing observations with “whole water column” data (PANGAEA.875579)</h2>
<p>Now we will fill missing observations with direct observations taken from the “whole water column.”</p>
<ol type="1">
<li>We will take bottle samples as the starting point (PANGAEA_875575)</li>
<li>If a bottle observation is missing fill with the closes CTD sensor observation (PANGAEA_875576)</li>
<li>If CTD sensor observation is missing fill with “whole water column” data (PANGAEA.875579). TBH I actually don’t understand where this data comes from. The PI listed is Sabrina Speich who is different than the most common PIs (Pesant and Guidi). But from what I can tell this is not the output of a model but somehow a direct observation corresponding to a Tara station.</li>
<li>Finally, if none of the three above exist we will fill the data with the output from the MIT Darwin model.</li>
</ol>
<p>This function creates a new variable replacing a variable from either PANGAEA_875575 or PANGAEA_875576 with the corresponding SRF/MIX/DCM/MES measurement from PANGAEA.875579</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>replace_w_875579 <span class="ot">&lt;-</span> <span class="cf">function</span>(.data, varname, target, srf, mix, dcm){</span>
<span id="cb7-2"><a href="#cb7-2"></a>  .data <span class="sc">%&gt;%</span> </span>
<span id="cb7-3"><a href="#cb7-3"></a>    <span class="fu">mutate</span>(<span class="st">"{varname}"</span> <span class="sc">:</span><span class="er">=</span> <span class="fu">if_else</span>(<span class="sc">!</span><span class="fu">is.na</span>({{ target }}), {{ target }},</span>
<span id="cb7-4"><a href="#cb7-4"></a>          <span class="fu">if_else</span>(<span class="fu">str_detect</span>(env_ontology, <span class="st">"SRF"</span>), {{ srf }},</span>
<span id="cb7-5"><a href="#cb7-5"></a>                  <span class="fu">if_else</span>(<span class="fu">str_detect</span>(env_ontology, <span class="st">"MIX"</span>), {{ mix }}, </span>
<span id="cb7-6"><a href="#cb7-6"></a>                          <span class="fu">if_else</span>(<span class="fu">str_detect</span>(env_ontology, <span class="st">"DCM"</span>), {{ dcm }}, <span class="cn">NA_real_</span>)))))</span>
<span id="cb7-7"><a href="#cb7-7"></a>}</span>
<span id="cb7-8"><a href="#cb7-8"></a></span>
<span id="cb7-9"><a href="#cb7-9"></a>simple_replace <span class="ot">&lt;-</span> <span class="cf">function</span>(.data, varname, target, replace){</span>
<span id="cb7-10"><a href="#cb7-10"></a>  .data <span class="sc">%&gt;%</span> </span>
<span id="cb7-11"><a href="#cb7-11"></a>    <span class="fu">mutate</span>(<span class="st">"{varname}"</span> <span class="sc">:</span><span class="er">=</span> <span class="fu">if_else</span>(<span class="fu">is.na</span>({{ target }}), {{ replace }}, {{ target }}))</span>
<span id="cb7-12"><a href="#cb7-12"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Call the above function for each of the 10 variables of interest</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>tara_biosamples_joined02_replaced <span class="ot">&lt;-</span> tara_biosamples_joined02 <span class="sc">%&gt;%</span></span>
<span id="cb8-2"><a href="#cb8-2"></a>  <span class="fu">replace_w_875579</span>(<span class="st">"density"</span>,     PANGAEA<span class="fl">.875576</span>_047, PANGAEA<span class="fl">.875579</span>_074, PANGAEA<span class="fl">.875579</span>_077, PANGAEA<span class="fl">.875579</span>_079) <span class="sc">%&gt;%</span> </span>
<span id="cb8-3"><a href="#cb8-3"></a>  <span class="fu">replace_w_875579</span>(<span class="st">"temperature"</span>, PANGAEA<span class="fl">.875576</span>_027, PANGAEA<span class="fl">.875579</span>_064, PANGAEA<span class="fl">.875579</span>_067, PANGAEA<span class="fl">.875579</span>_069) <span class="sc">%&gt;%</span> </span>
<span id="cb8-4"><a href="#cb8-4"></a>  <span class="fu">replace_w_875579</span>(<span class="st">"salinity"</span>,    PANGAEA<span class="fl">.875576</span>_037, PANGAEA<span class="fl">.875579</span>_054, PANGAEA<span class="fl">.875579</span>_057, PANGAEA<span class="fl">.875579</span>_059) <span class="sc">%&gt;%</span> </span>
<span id="cb8-5"><a href="#cb8-5"></a>  <span class="fu">replace_w_875579</span>(<span class="st">"chl_a"</span>,       PANGAEA<span class="fl">.875576</span>_077, PANGAEA<span class="fl">.875579</span>_094, PANGAEA<span class="fl">.875579</span>_097, PANGAEA<span class="fl">.875579</span>_099) <span class="sc">%&gt;%</span> </span>
<span id="cb8-6"><a href="#cb8-6"></a>  <span class="fu">replace_w_875579</span>(<span class="st">"dO2"</span>,         PANGAEA<span class="fl">.875576</span>_057, PANGAEA<span class="fl">.875579</span>_104, PANGAEA<span class="fl">.875579</span>_107, PANGAEA<span class="fl">.875579</span>_109) <span class="sc">%&gt;%</span> </span>
<span id="cb8-7"><a href="#cb8-7"></a>  <span class="fu">replace_w_875579</span>(<span class="st">"DIN"</span>,         PANGAEA<span class="fl">.875575</span>_031, PANGAEA<span class="fl">.875579</span>_114, PANGAEA<span class="fl">.875579</span>_117, PANGAEA<span class="fl">.875579</span>_119) <span class="sc">%&gt;%</span> </span>
<span id="cb8-8"><a href="#cb8-8"></a>    <span class="fu">simple_replace</span>(<span class="st">"DIN"</span>,         DIN, PANGAEA<span class="fl">.875577</span>_045) <span class="sc">%&gt;%</span> </span>
<span id="cb8-9"><a href="#cb8-9"></a>    <span class="fu">simple_replace</span>(<span class="st">"NO2"</span>,         PANGAEA<span class="fl">.875575</span>_021, PANGAEA<span class="fl">.875577</span>_043) <span class="sc">%&gt;%</span> </span>
<span id="cb8-10"><a href="#cb8-10"></a>  <span class="fu">mutate</span>(<span class="at">FeT =</span> PANGAEA<span class="fl">.875577</span>_039, </span>
<span id="cb8-11"><a href="#cb8-11"></a>         <span class="at">PO4 =</span> PANGAEA<span class="fl">.875575</span>_026,</span>
<span id="cb8-12"><a href="#cb8-12"></a>         <span class="at">SiO2 =</span> PANGAEA<span class="fl">.875575</span>_036) <span class="sc">%&gt;%</span> </span>
<span id="cb8-13"><a href="#cb8-13"></a>  <span class="fu">relocate</span>(tara_barcode_num, tara_station, env_ontology, latitude, longitude, date_time, depth, depth_min, depth_max, </span>
<span id="cb8-14"><a href="#cb8-14"></a>           density, temperature, salinity, chl_a, dO2, DIN, NO2, FeT, PO4, SiO2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Fix some stuff with dates</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>statistical_mode <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb9-2"><a href="#cb9-2"></a>  ux <span class="ot">&lt;-</span> <span class="fu">unique</span>(x)</span>
<span id="cb9-3"><a href="#cb9-3"></a>  ux[<span class="fu">which.max</span>(<span class="fu">tabulate</span>(<span class="fu">match</span>(x, ux)))]</span>
<span id="cb9-4"><a href="#cb9-4"></a>}</span>
<span id="cb9-5"><a href="#cb9-5"></a></span>
<span id="cb9-6"><a href="#cb9-6"></a>tara_biosamples_joined02_replaced <span class="ot">&lt;-</span> tara_biosamples_joined02_replaced <span class="sc">%&gt;%</span> </span>
<span id="cb9-7"><a href="#cb9-7"></a>  <span class="co"># convert the date_time to only date. This ensures that when you are grouping</span></span>
<span id="cb9-8"><a href="#cb9-8"></a>  <span class="co"># by tara_station and date_time that you don't expand the results to include</span></span>
<span id="cb9-9"><a href="#cb9-9"></a>  <span class="co"># multiple times within a single station</span></span>
<span id="cb9-10"><a href="#cb9-10"></a>  <span class="fu">mutate</span>(<span class="at">date_time =</span> lubridate<span class="sc">::</span><span class="fu">as_date</span>(stringr<span class="sc">::</span><span class="fu">str_extract</span>(date_time, <span class="st">"^</span><span class="sc">\\</span><span class="st">d{4}-</span><span class="sc">\\</span><span class="st">d{2}-</span><span class="sc">\\</span><span class="st">d{2}"</span>))) <span class="sc">%&gt;%</span> </span>
<span id="cb9-11"><a href="#cb9-11"></a>  <span class="fu">group_by</span>(tara_station) <span class="sc">%&gt;%</span> </span>
<span id="cb9-12"><a href="#cb9-12"></a>  <span class="co"># in the case that there are multiple days sampled at a station take the most</span></span>
<span id="cb9-13"><a href="#cb9-13"></a>  <span class="co"># common day to represent the whole station</span></span>
<span id="cb9-14"><a href="#cb9-14"></a>  <span class="fu">mutate</span>(<span class="at">date_time =</span> <span class="fu">statistical_mode</span>(date_time)) <span class="sc">%&gt;%</span> </span>
<span id="cb9-15"><a href="#cb9-15"></a>  <span class="fu">ungroup</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Now after we have filled observations with data internally available from the Tara Pangaea collection, we need to find which stations still have missing data observations for one or more these variables.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>target_vars <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"density"</span>, <span class="st">"temperature"</span>, <span class="st">"salinity"</span>, <span class="st">"chl_a"</span>, <span class="st">"dO2"</span>, <span class="st">"DIN"</span>, <span class="st">"NO2"</span>, <span class="st">"FeT"</span>, <span class="st">"PO4"</span>, <span class="st">"SiO2"</span>)</span>
<span id="cb10-2"><a href="#cb10-2"></a></span>
<span id="cb10-3"><a href="#cb10-3"></a>missing <span class="ot">&lt;-</span> tara_biosamples_joined02_replaced <span class="sc">%&gt;%</span> </span>
<span id="cb10-4"><a href="#cb10-4"></a>  <span class="fu">filter</span>(<span class="fu">if_any</span>(<span class="fu">all_of</span>(target_vars), is.na)) <span class="sc">%&gt;%</span></span>
<span id="cb10-5"><a href="#cb10-5"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="fu">all_of</span>(target_vars), <span class="sc">~</span><span class="fu">any</span>(<span class="fu">is.na</span>(.))),</span>
<span id="cb10-6"><a href="#cb10-6"></a>            <span class="at">.by =</span> <span class="fu">c</span>(tara_station, env_ontology, depth, depth_min, depth_max, latitude, longitude, date_time))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Most missing observations are because of PO4 and SiO2</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>missing <span class="sc">%&gt;%</span> </span>
<span id="cb11-2"><a href="#cb11-2"></a>  <span class="fu">summarize</span>(<span class="fu">across</span>(<span class="fu">all_of</span>(target_vars), sum))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["density"],"name":[1],"type":["int"],"align":["right"]},{"label":["temperature"],"name":[2],"type":["int"],"align":["right"]},{"label":["salinity"],"name":[3],"type":["int"],"align":["right"]},{"label":["chl_a"],"name":[4],"type":["int"],"align":["right"]},{"label":["dO2"],"name":[5],"type":["int"],"align":["right"]},{"label":["DIN"],"name":[6],"type":["int"],"align":["right"]},{"label":["NO2"],"name":[7],"type":["int"],"align":["right"]},{"label":["FeT"],"name":[8],"type":["int"],"align":["right"]},{"label":["PO4"],"name":[9],"type":["int"],"align":["right"]},{"label":["SiO2"],"name":[10],"type":["int"],"align":["right"]}],"data":[{"1":"1","2":"1","3":"1","4":"4","5":"17","6":"0","7":"0","8":"0","9":"66","10":"66"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>And this is for 56 different tara stations</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="fu">distinct</span>(missing, tara_station) <span class="sc">%&gt;%</span> </span>
<span id="cb12-2"><a href="#cb12-2"></a>  <span class="fu">count</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["n"],"name":[1],"type":["int"],"align":["right"]}],"data":[{"1":"56"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>This level of completion is pretty OK I think… To replace the missing fields I will use modeled variables from the MIT Darwin model. These can be obtained from <a href="https://simonscmap.com/">Simons CMAP.</a></p>
<p>Set the api-key for CMAP. This allows your R session to communicate with their backend database. You will need to get your own key following instructions <a href="https://simonscmap.com/apikeymanagement">here</a>.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="fu">library</span>(cmap4r)</span>
<span id="cb13-2"><a href="#cb13-2"></a>cmap4r<span class="sc">::</span><span class="fu">set_authorization</span>(<span class="at">reset =</span> <span class="cn">TRUE</span>)</span>
<span id="cb13-3"><a href="#cb13-3"></a>cmap4r<span class="sc">::</span><span class="fu">set_authorization</span>(<span class="at">cmap_key =</span> <span class="st">"KEY-GOES-HERE"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Download a local copy of the CMAP catalog. Useful for querying later…</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>local_cat <span class="ot">&lt;-</span> cmap4r<span class="sc">::</span><span class="fu">get_catalog</span>() <span class="sc">%&gt;%</span></span>
<span id="cb14-2"><a href="#cb14-2"></a>  <span class="fu">select</span>(Variable, Table_Name, Unit, Sensor, Unit)</span>
<span id="cb14-3"><a href="#cb14-3"></a></span>
<span id="cb14-4"><a href="#cb14-4"></a><span class="fu">write_tsv</span>(local_cat, here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"_data_raw"</span>, <span class="st">"cmap"</span>, <span class="st">"cmap_catalog.tsv"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="fill-missing-observations-with-darwin-model-nutrients-output" class="level2" data-number="3.4">
<h2 data-number="3.4" class="anchored" data-anchor-id="fill-missing-observations-with-darwin-model-nutrients-output"><span class="header-section-number">3.4</span> Fill missing observations with Darwin model Nutrients output</h2>
<section id="darwin-coordinates-for-nutrients" class="level3" data-number="3.4.1">
<h3 data-number="3.4.1" class="anchored" data-anchor-id="darwin-coordinates-for-nutrients"><span class="header-section-number">3.4.1</span> Darwin coordinates for nutrients</h3>
<p>CMAP has some select nutrient concentrations available from the <a href="https://simonscmap.com/catalog/datasets/Darwin_Nutrient">3-day averagedDarwin Model output</a>. These include</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>local_cat <span class="sc">%&gt;%</span> </span>
<span id="cb15-2"><a href="#cb15-2"></a>  <span class="fu">filter</span>(Table_Name <span class="sc">==</span> <span class="st">"tblDarwin_Nutrient"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["Variable"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Table_Name"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Unit"],"name":[3],"type":["chr"],"align":["left"]},{"label":["Sensor"],"name":[4],"type":["chr"],"align":["left"]}],"data":[{"1":"FeT","2":"tblDarwin_Nutrient","3":"mmol Fe","4":"Blend"},{"1":"PO4","2":"tblDarwin_Nutrient","3":"mmol PO4","4":"Blend"},{"1":"DIN","2":"tblDarwin_Nutrient","3":"mmol DIN","4":"Blend"},{"1":"SiO2","2":"tblDarwin_Nutrient","3":"mmol Si","4":"Blend"},{"1":"O2","2":"tblDarwin_Nutrient","3":"mmol O2","4":"Blend"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Not exactly clear what the units are for these nutrients, but my guess is mmol m-3 since most other concentration units are per 1 m3 unit volume.</p>
<p>Now we will get only the stations with missing nutrients in <code>tblDarwin_Nutrient</code></p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>missing_nuts <span class="ot">&lt;-</span> missing <span class="sc">%&gt;%</span> </span>
<span id="cb16-2"><a href="#cb16-2"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="fu">c</span>(<span class="sc">-</span>tara_station, <span class="sc">-</span>env_ontology, <span class="sc">-</span>depth, <span class="sc">-</span>depth_min, <span class="sc">-</span>depth_max, <span class="sc">-</span>latitude, <span class="sc">-</span>longitude, <span class="sc">-</span>date_time)) <span class="sc">%&gt;%</span> </span>
<span id="cb16-3"><a href="#cb16-3"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(value <span class="sc">==</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb16-4"><a href="#cb16-4"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(tara_station, env_ontology, latitude, longitude, date_time, depth, depth_min, depth_max) <span class="sc">%&gt;%</span> </span>
<span id="cb16-5"><a href="#cb16-5"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Now retrieve parameters from these stations that can be used to query CMAP database. We need a start latitude, ending latitude, starting longitude, ending longitude, starting depth, ending depth, and starting date and ending date. Note: here we don’t bother with depths because when you query all the variables in the CMAP database it returns all depths regardless of what you provide to the function (bug? feature?)</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>coord_range <span class="ot">&lt;-</span> <span class="cf">function</span>(x, f, offset, <span class="at">negate =</span> <span class="cn">FALSE</span>){</span>
<span id="cb17-2"><a href="#cb17-2"></a>  <span class="fu">return</span>(<span class="fu">if_else</span>(negate, <span class="sc">-</span>(<span class="fu">f</span>(<span class="fu">max</span>(<span class="fu">abs</span>(x))) <span class="sc">+</span> offset), <span class="fu">f</span>(<span class="fu">max</span>(<span class="fu">abs</span>(x))) <span class="sc">+</span> offset))</span>
<span id="cb17-3"><a href="#cb17-3"></a>}</span>
<span id="cb17-4"><a href="#cb17-4"></a></span>
<span id="cb17-5"><a href="#cb17-5"></a>missing_nuts_nested <span class="ot">&lt;-</span> missing_nuts <span class="sc">%&gt;%</span> </span>
<span id="cb17-6"><a href="#cb17-6"></a>  <span class="fu">group_by</span>(tara_station) <span class="sc">%&gt;%</span> </span>
<span id="cb17-7"><a href="#cb17-7"></a>  <span class="fu">mutate</span>(<span class="at">lat1 =</span> <span class="fu">if_else</span>(latitude <span class="sc">&lt;</span> <span class="dv">0</span>,   <span class="fu">coord_range</span>(latitude, ceiling, <span class="dv">1</span>, <span class="cn">TRUE</span>),  <span class="fu">coord_range</span>(latitude, floor, <span class="sc">-</span><span class="dv">1</span>, <span class="cn">FALSE</span>)),</span>
<span id="cb17-8"><a href="#cb17-8"></a>         <span class="at">lat2 =</span> <span class="fu">if_else</span>(latitude <span class="sc">&lt;</span> <span class="dv">0</span>,   <span class="fu">coord_range</span>(latitude, floor, <span class="sc">-</span><span class="dv">1</span>, <span class="cn">TRUE</span>),   <span class="fu">coord_range</span>(latitude, ceiling, <span class="dv">1</span>, <span class="cn">FALSE</span>)),</span>
<span id="cb17-9"><a href="#cb17-9"></a>         <span class="at">lon1 =</span> <span class="fu">if_else</span>(longitude <span class="sc">&lt;</span> <span class="dv">0</span>,  <span class="fu">coord_range</span>(longitude, ceiling, <span class="dv">1</span>, <span class="cn">TRUE</span>),  <span class="fu">coord_range</span>(longitude, floor, <span class="sc">-</span><span class="dv">1</span>, <span class="cn">FALSE</span>)),</span>
<span id="cb17-10"><a href="#cb17-10"></a>         <span class="at">lon2 =</span> <span class="fu">if_else</span>(longitude <span class="sc">&lt;</span> <span class="dv">0</span>,  <span class="fu">coord_range</span>(longitude, floor, <span class="sc">-</span><span class="dv">1</span>, <span class="cn">TRUE</span>), <span class="fu">coord_range</span>(longitude, ceiling, <span class="dv">1</span>, <span class="cn">FALSE</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb17-11"><a href="#cb17-11"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">dt1 =</span> <span class="fu">floor_date</span>(date_time, <span class="st">'month'</span>),</span>
<span id="cb17-12"><a href="#cb17-12"></a>                <span class="at">dt2 =</span> <span class="fu">ceiling_date</span>(date_time, <span class="st">'month'</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb17-13"><a href="#cb17-13"></a>  <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb17-14"><a href="#cb17-14"></a>  <span class="fu">nest</span>(<span class="at">original_data =</span> <span class="fu">c</span>(env_ontology<span class="sc">:</span>depth_max)) <span class="sc">%&gt;%</span> </span>
<span id="cb17-15"><a href="#cb17-15"></a>  <span class="fu">nest</span>(<span class="at">dw_query =</span> <span class="fu">c</span>(lat1<span class="sc">:</span>dt2))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Now make the call querying the CMAP remote database. This maps a separate call to every unique missing Tara Station. Even though this isn’t very efficient it is faster and takes less space that just downloaded the Darwin output for the entire ocean and then later filtering to the coordinates of interest.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a>missing_nuts_nested_darwin <span class="ot">&lt;-</span> missing_nuts_nested <span class="sc">%&gt;%</span> </span>
<span id="cb18-2"><a href="#cb18-2"></a>  <span class="fu">mutate</span>(</span>
<span id="cb18-3"><a href="#cb18-3"></a>    <span class="at">tbl =</span> <span class="fu">map</span>(</span>
<span id="cb18-4"><a href="#cb18-4"></a>      dw_query,</span>
<span id="cb18-5"><a href="#cb18-5"></a>      <span class="cf">function</span>(df) cmap4r<span class="sc">::</span><span class="fu">get_spacetime</span>(</span>
<span id="cb18-6"><a href="#cb18-6"></a>        <span class="at">tableName =</span> <span class="st">'tblDarwin_Nutrient'</span>,</span>
<span id="cb18-7"><a href="#cb18-7"></a>        <span class="at">varName =</span> <span class="st">'*'</span>,</span>
<span id="cb18-8"><a href="#cb18-8"></a>        <span class="at">dt1 =</span> <span class="fu">as.character</span>(df<span class="sc">$</span>dt1),</span>
<span id="cb18-9"><a href="#cb18-9"></a>        <span class="at">dt2 =</span> <span class="fu">as.character</span>(df<span class="sc">$</span>dt2),</span>
<span id="cb18-10"><a href="#cb18-10"></a>        <span class="at">lat1 =</span> df<span class="sc">$</span>lat1,</span>
<span id="cb18-11"><a href="#cb18-11"></a>        <span class="at">lat2 =</span> df<span class="sc">$</span>lat2,</span>
<span id="cb18-12"><a href="#cb18-12"></a>        <span class="at">lon1 =</span> df<span class="sc">$</span>lon1,</span>
<span id="cb18-13"><a href="#cb18-13"></a>        <span class="at">lon2 =</span> df<span class="sc">$</span>lon2)))</span>
<span id="cb18-14"><a href="#cb18-14"></a></span>
<span id="cb18-15"><a href="#cb18-15"></a><span class="co"># save this for later to speed up quarto renders</span></span>
<span id="cb18-16"><a href="#cb18-16"></a><span class="fu">write_rds</span>(missing_nuts_nested_darwin, here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"_data_raw"</span>, <span class="st">"tara"</span>, <span class="st">"cmap_missing_nuts.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="filter-darwin-output-to-tara-coordinates" class="level3" data-number="3.4.2">
<h3 data-number="3.4.2" class="anchored" data-anchor-id="filter-darwin-output-to-tara-coordinates"><span class="header-section-number">3.4.2</span> Filter Darwin output to Tara coordinates</h3>
<p>Now we need to whittle down the Darwin model output to only the time, depth, and coordinates of interest. We do this by</p>
<ol type="1">
<li>Fuzzy joining on latitude and longitude using <code>fuzzyjoin::geo_inner_join</code>. Briefly this calculates the <a href="https://en.wikipedia.org/wiki/Haversine_formula">Haversine distance</a> between two sets of coordinates and keeps only those with distance magnitude less than <code>max_dist</code> (here we set this to 65 kilometers). This allows us to keep only coordinates that are as close as possible to the sampled Tara coordinates. Later we select for the smallest Haversine distance to select a point from the model grid that is closest to the Tara station. Sometimes these coordinates will be a land cell on the Darwin grid and will return NAs so we use <code>drop_na</code> to only include coordinates in the ocean with nutrients.</li>
<li>Next we calculate the duration of time between the Tara observation and the time step from the Darwin Model. Note this is the 3-day averaged model output. We then filter to only include model output that is closest in time (shortest duration) to the Tara observation.</li>
<li>Finally, we need to make a measure of the distance between the sampled Tara depth and the model output depth. In the upper 250 the depth is high resolution so usually we find a depth between <code>depth_min</code> and <code>depth_max</code> in the model output. Deeper in the ocean the model output is only every 100 meters or more so we need to filter on the smallest possible difference from the Tara depth</li>
</ol>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>missing_nuts_nested_darwin_filtered <span class="ot">&lt;-</span> missing_nuts_nested_darwin <span class="sc">%&gt;%</span> </span>
<span id="cb19-2"><a href="#cb19-2"></a>  <span class="co"># unnest on the Darwin output. expand rows per Tara Station</span></span>
<span id="cb19-3"><a href="#cb19-3"></a>  <span class="fu">unnest</span>(tbl) <span class="sc">%&gt;%</span></span>
<span id="cb19-4"><a href="#cb19-4"></a>  <span class="co"># select only relevant columns to make management easier</span></span>
<span id="cb19-5"><a href="#cb19-5"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(tara_station, </span>
<span id="cb19-6"><a href="#cb19-6"></a>                <span class="at">date_time =</span> time, <span class="at">latitude =</span> lat, <span class="at">longitude =</span> lon, depth,</span>
<span id="cb19-7"><a href="#cb19-7"></a>                <span class="at">PO4_dw =</span> PO4, <span class="at">SiO2_dw =</span> SiO2, <span class="at">O2_dw =</span> O2) <span class="sc">%&gt;%</span></span>
<span id="cb19-8"><a href="#cb19-8"></a>  <span class="co"># fuzzy join the Darwin output to the coordinates of missing data using the </span></span>
<span id="cb19-9"><a href="#cb19-9"></a>  <span class="co"># haversine distance and excluding joins that exceed 65 km in distance apart</span></span>
<span id="cb19-10"><a href="#cb19-10"></a>  fuzzyjoin<span class="sc">::</span><span class="fu">geo_inner_join</span>(missing_nuts,</span>
<span id="cb19-11"><a href="#cb19-11"></a>                            <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"latitude"</span>, <span class="st">"longitude"</span>),</span>
<span id="cb19-12"><a href="#cb19-12"></a>                            <span class="at">distance_col =</span> <span class="st">"geojoin_dist_km"</span>,</span>
<span id="cb19-13"><a href="#cb19-13"></a>                            <span class="at">max_dist =</span> <span class="dv">65</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb19-14"><a href="#cb19-14"></a>  <span class="co"># narrow down the data so we only consider matching Tara stations from the two datasets</span></span>
<span id="cb19-15"><a href="#cb19-15"></a>  <span class="fu">filter</span>(tara_station.x <span class="sc">==</span> tara_station.y) <span class="sc">%&gt;%</span></span>
<span id="cb19-16"><a href="#cb19-16"></a>  <span class="co"># exclude any Darwin observations that have NAs for PO4, SiO2, or O2 (these are genarally land grids or seafloor)</span></span>
<span id="cb19-17"><a href="#cb19-17"></a>  <span class="fu">filter</span>(<span class="sc">!</span>(<span class="fu">is.na</span>(PO4_dw) <span class="sc">&amp;</span> <span class="fu">is.na</span>(SiO2_dw) <span class="sc">&amp;</span> <span class="fu">is.na</span>(O2_dw))) <span class="sc">%&gt;%</span> </span>
<span id="cb19-18"><a href="#cb19-18"></a>  <span class="co"># Group by unique variables for each Tara observation</span></span>
<span id="cb19-19"><a href="#cb19-19"></a>  <span class="fu">group_by</span>(tara_station.y, env_ontology, depth.y, depth_min, depth_max) <span class="sc">%&gt;%</span> </span>
<span id="cb19-20"><a href="#cb19-20"></a>  <span class="co"># Filter to include only matches with smallest distance between Darwin grid and the coords of the direct observation</span></span>
<span id="cb19-21"><a href="#cb19-21"></a>  <span class="fu">filter</span>(geojoin_dist_km <span class="sc">==</span> <span class="fu">min</span>(geojoin_dist_km)) <span class="sc">%&gt;%</span> </span>
<span id="cb19-22"><a href="#cb19-22"></a>  <span class="co"># do some type conversions for dates to allow subsequent calculations</span></span>
<span id="cb19-23"><a href="#cb19-23"></a>  <span class="fu">mutate</span>(<span class="at">date_time.x =</span> lubridate<span class="sc">::</span><span class="fu">as_date</span>(date_time.x),</span>
<span id="cb19-24"><a href="#cb19-24"></a>         <span class="at">date_time.y =</span> lubridate<span class="sc">::</span><span class="fu">as_date</span>(stringr<span class="sc">::</span><span class="fu">str_extract</span>(date_time.y, <span class="st">"^</span><span class="sc">\\</span><span class="st">d{4}-</span><span class="sc">\\</span><span class="st">d{2}-</span><span class="sc">\\</span><span class="st">d{2}"</span>))) <span class="sc">%&gt;%</span> </span>
<span id="cb19-25"><a href="#cb19-25"></a>  <span class="co"># calculate the difference between the model output (3-day interval) and the sampling time of the direct observation</span></span>
<span id="cb19-26"><a href="#cb19-26"></a>  <span class="fu">mutate</span>(<span class="at">time_diff =</span> <span class="fu">as.duration</span>(<span class="fu">abs</span>(<span class="fu">interval</span>(date_time.x, date_time.y)))) <span class="sc">%&gt;%</span> </span>
<span id="cb19-27"><a href="#cb19-27"></a>  <span class="co"># filter to only include matches with the smallest time difference</span></span>
<span id="cb19-28"><a href="#cb19-28"></a>  <span class="fu">filter</span>(time_diff <span class="sc">==</span> <span class="fu">min</span>(time_diff)) <span class="sc">%&gt;%</span> </span>
<span id="cb19-29"><a href="#cb19-29"></a>  <span class="co"># filter to only include matches with the smallest depth difference</span></span>
<span id="cb19-30"><a href="#cb19-30"></a>  <span class="fu">filter</span>(<span class="fu">abs</span>(depth.y <span class="sc">-</span> depth.x) <span class="sc">==</span> <span class="fu">min</span>(<span class="fu">abs</span>(depth.y <span class="sc">-</span> depth.x))) <span class="sc">%&gt;%</span> </span>
<span id="cb19-31"><a href="#cb19-31"></a>  <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb19-32"><a href="#cb19-32"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="at">tara_station =</span> tara_station.y, </span>
<span id="cb19-33"><a href="#cb19-33"></a>                <span class="at">depth =</span> depth.y, depth_min, depth_max, env_ontology,</span>
<span id="cb19-34"><a href="#cb19-34"></a>                <span class="at">latitude =</span> latitude.y, <span class="at">longitude =</span> longitude.y, PO4_dw, SiO2_dw, O2_dw) <span class="sc">%&gt;%</span></span>
<span id="cb19-35"><a href="#cb19-35"></a>  <span class="co"># take mean of multi-observations (e.g. if the real depth is 4, the closest</span></span>
<span id="cb19-36"><a href="#cb19-36"></a>  <span class="co"># depths from Darwin are 35 and 45 so we average the concentration from the</span></span>
<span id="cb19-37"><a href="#cb19-37"></a>  <span class="co"># two flanking depths)</span></span>
<span id="cb19-38"><a href="#cb19-38"></a>  <span class="fu">summarize</span>(<span class="fu">across</span>(PO4_dw<span class="sc">:</span>O2_dw, mean), </span>
<span id="cb19-39"><a href="#cb19-39"></a>            <span class="at">.by =</span> <span class="fu">c</span>(tara_station<span class="sc">:</span>longitude))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="join-observed-and-darwin-model-output-into-single-dataframe" class="level3" data-number="3.4.3">
<h3 data-number="3.4.3" class="anchored" data-anchor-id="join-observed-and-darwin-model-output-into-single-dataframe"><span class="header-section-number">3.4.3</span> Join observed and Darwin Model output into single dataframe</h3>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>tara_biosamples_joined02_replaced02 <span class="ot">&lt;-</span> <span class="fu">left_join</span>(tara_biosamples_joined02_replaced, missing_nuts_nested_darwin_filtered, </span>
<span id="cb20-2"><a href="#cb20-2"></a>          <span class="at">by =</span> <span class="fu">join_by</span>(tara_station, env_ontology, latitude, longitude, depth, depth_min, depth_max)) <span class="sc">%&gt;%</span> </span>
<span id="cb20-3"><a href="#cb20-3"></a>  <span class="fu">simple_replace</span>(<span class="st">"SiO2"</span>, SiO2, SiO2_dw) <span class="sc">%&gt;%</span></span>
<span id="cb20-4"><a href="#cb20-4"></a>  <span class="fu">simple_replace</span>(<span class="st">"dO2"</span>,  dO2, O2_dw) <span class="sc">%&gt;%</span></span>
<span id="cb20-5"><a href="#cb20-5"></a>  <span class="fu">simple_replace</span>(<span class="st">"PO4"</span>,  PO4, PO4_dw)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="impute-remaining-missing-observations" class="level2" data-number="3.5">
<h2 data-number="3.5" class="anchored" data-anchor-id="impute-remaining-missing-observations"><span class="header-section-number">3.5</span> Impute remaining missing observations</h2>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="fu">library</span>(missRanger)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>There are three Tara stations with missing observations. <code>TARA_011</code> is missing chl_a, density, salinity and temperature, while the others are missing only chl_a</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a>missing <span class="sc">%&gt;%</span> </span>
<span id="cb22-2"><a href="#cb22-2"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(tara_station<span class="sc">:</span>chl_a) <span class="sc">%&gt;%</span></span>
<span id="cb22-3"><a href="#cb22-3"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="fu">c</span>(<span class="sc">-</span>tara_station, <span class="sc">-</span>env_ontology, <span class="sc">-</span>depth, <span class="sc">-</span>depth_min, <span class="sc">-</span>depth_max, <span class="sc">-</span>latitude, <span class="sc">-</span>longitude, <span class="sc">-</span>date_time)) <span class="sc">%&gt;%</span> </span>
<span id="cb22-4"><a href="#cb22-4"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(value <span class="sc">==</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb22-5"><a href="#cb22-5"></a>  <span class="fu">group_by</span>(tara_station, name) <span class="sc">%&gt;%</span> </span>
<span id="cb22-6"><a href="#cb22-6"></a>  <span class="fu">count</span>(value) <span class="sc">%&gt;%</span> </span>
<span id="cb22-7"><a href="#cb22-7"></a>  <span class="fu">arrange</span>(name)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["tara_station"],"name":[1],"type":["chr"],"align":["left"]},{"label":["name"],"name":[2],"type":["chr"],"align":["left"]},{"label":["value"],"name":[3],"type":["lgl"],"align":["right"]},{"label":["n"],"name":[4],"type":["int"],"align":["right"]}],"data":[{"1":"TARA_011","2":"chl_a","3":"TRUE","4":"1"},{"1":"TARA_012","2":"chl_a","3":"TRUE","4":"2"},{"1":"TARA_049","2":"chl_a","3":"TRUE","4":"1"},{"1":"TARA_011","2":"density","3":"TRUE","4":"1"},{"1":"TARA_011","2":"salinity","3":"TRUE","4":"1"},{"1":"TARA_011","2":"temperature","3":"TRUE","4":"1"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Since this is only a few missing observations, we are going to impute them using the <a href="https://mayer79.github.io/missRanger/index.html"><code>missRanger</code> package</a>, which uses the <a href="https://doi.org/10.1093/bioinformatics/btr597"><code>missForest</code> random forest imputation approach.</a></p>
<p>First only subset variables of interest and make the data unique</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>tara_biosamples_joined02_replaced02_uniq <span class="ot">&lt;-</span> tara_biosamples_joined02_replaced02 <span class="sc">%&gt;%</span> </span>
<span id="cb23-2"><a href="#cb23-2"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(tara_station<span class="sc">:</span>SiO2) <span class="sc">%&gt;%</span> </span>
<span id="cb23-3"><a href="#cb23-3"></a>  <span class="fu">distinct</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Run the missRanger algorithm to impute missing values. We impute density, temperature, salinity, chl_a by depth, density, temperature, salinity, chl_a, dO2, DIN, NO2, FeT, PO4, SiO2</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="co"># set seed for reproducibility</span></span>
<span id="cb24-2"><a href="#cb24-2"></a><span class="fu">set.seed</span>(<span class="dv">35782</span>)</span>
<span id="cb24-3"><a href="#cb24-3"></a>tara_biosamples_joined02_replaced02_uniq_rangerd <span class="ot">&lt;-</span> missRanger<span class="sc">::</span><span class="fu">missRanger</span>(tara_biosamples_joined02_replaced02_uniq, </span>
<span id="cb24-4"><a href="#cb24-4"></a>                                                                           . <span class="sc">~</span> . <span class="sc">-</span> tara_station <span class="sc">-</span> env_ontology <span class="sc">-</span> latitude <span class="sc">-</span> longitude <span class="sc">-</span> date_time <span class="sc">-</span> depth_min <span class="sc">-</span> depth_max, <span class="at">pmm.k =</span> <span class="dv">10</span>, <span class="at">num.trees =</span> <span class="dv">500</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="joining-and-replacing-missing-values-with-those-imputed" class="level2" data-number="3.6">
<h2 data-number="3.6" class="anchored" data-anchor-id="joining-and-replacing-missing-values-with-those-imputed"><span class="header-section-number">3.6</span> Joining and replacing missing values with those imputed</h2>
<p>First get only the samples that needed salinity, temp, density, chl_a replaced from the imputed dataset</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a>tara_biosamples_joined02_replaced02_uniq_imputed <span class="ot">&lt;-</span> tara_biosamples_joined02_replaced02_uniq <span class="sc">%&gt;%</span> <span class="fu">filter</span>(<span class="fu">if_any</span>(<span class="fu">everything</span>(), is.na)) <span class="sc">%&gt;%</span> <span class="fu">select</span>(tara_station<span class="sc">:</span>depth_max)</span>
<span id="cb25-2"><a href="#cb25-2"></a></span>
<span id="cb25-3"><a href="#cb25-3"></a>tara_biosamples_envdata_01 <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(tara_biosamples_joined02_replaced02_uniq_rangerd, tara_biosamples_joined02_replaced02_uniq_imputed,</span>
<span id="cb25-4"><a href="#cb25-4"></a>           <span class="at">by =</span> <span class="fu">join_by</span>(tara_station, env_ontology, latitude, longitude, date_time, depth, depth_min, depth_max))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Next join those imputed samples to the full <code>tara_biosamples_joined02_replaced02</code> table</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a>tara_biosamples_envdata_02 <span class="ot">&lt;-</span> <span class="fu">right_join</span>(tara_biosamples_envdata_01, tara_biosamples_joined02_replaced02,</span>
<span id="cb26-2"><a href="#cb26-2"></a>              <span class="at">by =</span> <span class="fu">join_by</span>(tara_station, env_ontology, latitude, longitude, date_time, depth, depth_min, depth_max))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="detour-to-add-some-additional-geographic-info" class="level3" data-number="3.6.1">
<h3 data-number="3.6.1" class="anchored" data-anchor-id="detour-to-add-some-additional-geographic-info"><span class="header-section-number">3.6.1</span> Detour to add some additional geographic info</h3>
<p>Next we want to add some geographic information to the dataset like longhurst code, ocean identifier, and the distance to the coast. We will calculate the distance to the nearest coast using <a href="https://mikkovihtakari.github.io/ggOceanMaps/reference/dist2land.html"><code>ggOceanMaps::dist2land</code></a>. <a href="https://blogs.ubc.ca/yiwang28/2024/01/06/my-r-learning-notes-3-how-to-calculate-the-shortest-distance-to-coast/">Inspired by this blogpost</a></p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a><span class="fu">library</span>(ggOceanMaps)</span>
<span id="cb27-2"><a href="#cb27-2"></a><span class="fu">library</span>(rnaturalearth)</span>
<span id="cb27-3"><a href="#cb27-3"></a><span class="fu">library</span>(sf)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Load coastline data at the largest most coarse scale (110) and also subset <code>PANGAEA_842237</code> to only coordinates and station identifier</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a>coast <span class="ot">&lt;-</span> rnaturalearth<span class="sc">::</span><span class="fu">ne_coastline</span>(<span class="at">scale=</span><span class="dv">110</span>, <span class="at">returnclass =</span> <span class="st">"sf"</span>)</span>
<span id="cb28-2"><a href="#cb28-2"></a>data <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">select</span>(PANGAEA_842237, tara_station, latitude, longitude, )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Plot to check</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a><span class="fu">plot</span>(coast[<span class="st">'featurecla'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="03_harmonize_ncbi_pangaea_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
<p>This function calculates great circle spherical distances (in kilometers) from a coordinate in the ocean to the nearest coastline using the <a href="https://r-spatial.github.io/sf/reference/geos_measures.html"><code>st_distance</code></a> function.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a>dist2land_df <span class="ot">&lt;-</span> ggOceanMaps<span class="sc">::</span><span class="fu">dist2land</span>(data, <span class="at">shapefile =</span> coast) <span class="sc">%&gt;%</span> </span>
<span id="cb30-2"><a href="#cb30-2"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(tara_station, <span class="at">dist_to_coastline =</span> ldist)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Used longitude and latitude as input coordinate column names in data</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Using custom land shapes.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating distances...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Returning great circle spherical distances from land as kilometers.</code></pre>
</div>
</div>
<p>Join the distance to coastline to other general large scale features like longhurst provinces</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a>ocean_feats_df <span class="ot">&lt;-</span> PANGAEA_842237 <span class="sc">%&gt;%</span></span>
<span id="cb35-2"><a href="#cb35-2"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(tara_station, <span class="at">ocean_id =</span> PANGAEA<span class="fl">.842237</span>_016, <span class="at">longhurst_biome =</span> PANGAEA<span class="fl">.842237</span>_015, </span>
<span id="cb35-3"><a href="#cb35-3"></a>                <span class="at">longhurst_code =</span> PANGAEA<span class="fl">.842237</span>_018, <span class="at">bathy_depth =</span> PANGAEA<span class="fl">.842237</span>_012) <span class="sc">%&gt;%</span> </span>
<span id="cb35-4"><a href="#cb35-4"></a>  <span class="fu">left_join</span>(dist2land_df, <span class="at">by =</span> <span class="fu">join_by</span>(tara_station))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="finalize" class="level3" data-number="3.6.2">
<h3 data-number="3.6.2" class="anchored" data-anchor-id="finalize"><span class="header-section-number">3.6.2</span> Finalize</h3>
<p>Now do the final joining to the full imputed data set, replace missing values with the imputed values, and select key variables</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a>tara_biosamples_envdata_final <span class="ot">&lt;-</span> <span class="fu">left_join</span>(tara_biosamples_envdata_02, ocean_feats_df, <span class="at">by =</span> <span class="fu">join_by</span>(tara_station)) <span class="sc">%&gt;%</span> </span>
<span id="cb36-2"><a href="#cb36-2"></a>  <span class="fu">simple_replace</span>(<span class="st">"density"</span>, density.y, density.x) <span class="sc">%&gt;%</span> </span>
<span id="cb36-3"><a href="#cb36-3"></a>  <span class="fu">simple_replace</span>(<span class="st">"salinity"</span>, salinity.y, salinity.x) <span class="sc">%&gt;%</span></span>
<span id="cb36-4"><a href="#cb36-4"></a>  <span class="fu">simple_replace</span>(<span class="st">"temperature"</span>, temperature.y, temperature.x) <span class="sc">%&gt;%</span></span>
<span id="cb36-5"><a href="#cb36-5"></a>  <span class="fu">simple_replace</span>(<span class="st">"chl_a"</span>, chl_a.y, chl_a.x) <span class="sc">%&gt;%</span> </span>
<span id="cb36-6"><a href="#cb36-6"></a>  <span class="fu">select</span>(tara_barcode_num, sra_acc_num<span class="sc">:</span>ena_acc_num, tara_station<span class="sc">:</span>date_time, size_low_thresh, size_high_thresh,</span>
<span id="cb36-7"><a href="#cb36-7"></a>         ocean_id, longhurst_code, longhurst_biome, bathy_depth, dist_to_coastline,</span>
<span id="cb36-8"><a href="#cb36-8"></a>         depth<span class="sc">:</span>depth_max, temperature, salinity, density, chl_a,</span>
<span id="cb36-9"><a href="#cb36-9"></a>         <span class="at">dO2 =</span> dO2.y, <span class="at">DIN =</span> DIN.y, <span class="at">NO2 =</span> NO2.y, <span class="at">FeT =</span> FeT.y, <span class="at">PO4 =</span> PO4.y, <span class="at">SiO2 =</span> SiO2.y)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Final check to see if there is missing entries in this dataframe - good to go</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a>tara_biosamples_envdata_final <span class="sc">%&gt;%</span> </span>
<span id="cb37-2"><a href="#cb37-2"></a>  <span class="fu">filter</span>(<span class="fu">if_any</span>(<span class="fu">everything</span>(), is.na))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["tara_barcode_num"],"name":[1],"type":["chr"],"align":["left"]},{"label":["sra_acc_num"],"name":[2],"type":["chr"],"align":["left"]},{"label":["sra_biosample_acc_num"],"name":[3],"type":["chr"],"align":["left"]},{"label":["sra_run_acc_num"],"name":[4],"type":["chr"],"align":["left"]},{"label":["sra_exp_acc_num"],"name":[5],"type":["chr"],"align":["left"]},{"label":["sra_study_acc_num"],"name":[6],"type":["chr"],"align":["left"]},{"label":["biosample_acc_num"],"name":[7],"type":["chr"],"align":["left"]},{"label":["ena_acc_num"],"name":[8],"type":["chr"],"align":["left"]},{"label":["tara_station"],"name":[9],"type":["chr"],"align":["left"]},{"label":["env_ontology"],"name":[10],"type":["chr"],"align":["left"]},{"label":["latitude"],"name":[11],"type":["dbl"],"align":["right"]},{"label":["longitude"],"name":[12],"type":["dbl"],"align":["right"]},{"label":["date_time"],"name":[13],"type":["date"],"align":["right"]},{"label":["size_low_thresh"],"name":[14],"type":["chr"],"align":["left"]},{"label":["size_high_thresh"],"name":[15],"type":["chr"],"align":["left"]},{"label":["ocean_id"],"name":[16],"type":["chr"],"align":["left"]},{"label":["longhurst_code"],"name":[17],"type":["chr"],"align":["left"]},{"label":["longhurst_biome"],"name":[18],"type":["chr"],"align":["left"]},{"label":["bathy_depth"],"name":[19],"type":["int"],"align":["right"]},{"label":["dist_to_coastline"],"name":[20],"type":["dbl"],"align":["right"]},{"label":["depth"],"name":[21],"type":["dbl"],"align":["right"]},{"label":["depth_min"],"name":[22],"type":["dbl"],"align":["right"]},{"label":["depth_max"],"name":[23],"type":["dbl"],"align":["right"]},{"label":["temperature"],"name":[24],"type":["dbl"],"align":["right"]},{"label":["salinity"],"name":[25],"type":["dbl"],"align":["right"]},{"label":["density"],"name":[26],"type":["dbl"],"align":["right"]},{"label":["chl_a"],"name":[27],"type":["dbl"],"align":["right"]},{"label":["dO2"],"name":[28],"type":["dbl"],"align":["right"]},{"label":["DIN"],"name":[29],"type":["dbl"],"align":["right"]},{"label":["NO2"],"name":[30],"type":["dbl"],"align":["right"]},{"label":["FeT"],"name":[31],"type":["dbl"],"align":["right"]},{"label":["PO4"],"name":[32],"type":["dbl"],"align":["right"]},{"label":["SiO2"],"name":[33],"type":["dbl"],"align":["right"]}],"data":[],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Save for later</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Show/hide code</summary>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a><span class="fu">write_tsv</span>(tara_biosamples_envdata_final, here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data"</span>, <span class="st">"tara"</span>, <span class="st">"tara_biosamples_envdata_final.tsv"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>


<!-- -->

</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb39" data-shortcodes="false"><pre class="sourceCode numberSource markdown number-lines code-with-copy"><code class="sourceCode markdown"><span id="cb39-1"><a href="#cb39-1"></a><span class="co">---</span></span>
<span id="cb39-2"><a href="#cb39-2"></a><span class="an">title:</span><span class="co"> "Harmonize environmental data and sequencing identifiers"</span></span>
<span id="cb39-3"><a href="#cb39-3"></a><span class="an">subtitle:</span><span class="co"> "Tara Oceans workflow"</span></span>
<span id="cb39-4"><a href="#cb39-4"></a><span class="an">author:</span><span class="co"> "Shane Hogle"</span></span>
<span id="cb39-5"><a href="#cb39-5"></a><span class="an">date:</span><span class="co"> today</span></span>
<span id="cb39-6"><a href="#cb39-6"></a><span class="an">abstract:</span><span class="co"> "Code in this notebook collates environmental parameters from different Pangaea.de datasets with the NCBI sequencing datasets. It then fills missing values from Pangaea with Darwin Model output for the nearest coordinate/depth/date. Finally, a small number of observations are imputed. The goal is to be able to associate an NCBI sequencing run accession number with environmental data for later statistical analysis. We do not want any missing data so we must try and patch missing data from other sources."</span></span>
<span id="cb39-7"><a href="#cb39-7"></a><span class="co">---</span></span>
<span id="cb39-8"><a href="#cb39-8"></a></span>
<span id="cb39-9"><a href="#cb39-9"></a><span class="fu"># Setup</span></span>
<span id="cb39-10"><a href="#cb39-10"></a></span>
<span id="cb39-11"><a href="#cb39-11"></a>This code loads required libraries and sets global variables</span>
<span id="cb39-12"><a href="#cb39-12"></a></span>
<span id="cb39-15"><a href="#cb39-15"></a><span class="in">```{r}</span></span>
<span id="cb39-16"><a href="#cb39-16"></a><span class="co">#| output: false</span></span>
<span id="cb39-17"><a href="#cb39-17"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb39-18"><a href="#cb39-18"></a><span class="fu">library</span>(here)</span>
<span id="cb39-19"><a href="#cb39-19"></a><span class="fu">library</span>(lubridate)</span>
<span id="cb39-20"><a href="#cb39-20"></a><span class="fu">library</span>(fuzzyjoin)</span>
<span id="cb39-21"><a href="#cb39-21"></a><span class="in">```</span></span>
<span id="cb39-22"><a href="#cb39-22"></a></span>
<span id="cb39-23"><a href="#cb39-23"></a><span class="fu"># Read data</span></span>
<span id="cb39-24"><a href="#cb39-24"></a></span>
<span id="cb39-25"><a href="#cb39-25"></a>Read NCBI BioSample data with unique Tara Ocean Barcodes </span>
<span id="cb39-26"><a href="#cb39-26"></a></span>
<span id="cb39-29"><a href="#cb39-29"></a><span class="in">```{r}</span></span>
<span id="cb39-30"><a href="#cb39-30"></a><span class="co">#| output: false</span></span>
<span id="cb39-31"><a href="#cb39-31"></a>tara_biosamples <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_tsv</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data"</span>, <span class="st">"tara"</span>, <span class="st">"ncbi2tara_barcode.tsv"</span>))</span>
<span id="cb39-32"><a href="#cb39-32"></a><span class="in">```</span></span>
<span id="cb39-33"><a href="#cb39-33"></a></span>
<span id="cb39-34"><a href="#cb39-34"></a>Read previously saved environmental data and metadata downloaded from Pangaea.de</span>
<span id="cb39-35"><a href="#cb39-35"></a></span>
<span id="cb39-38"><a href="#cb39-38"></a><span class="in">```{r}</span></span>
<span id="cb39-39"><a href="#cb39-39"></a><span class="co"># tara stations</span></span>
<span id="cb39-40"><a href="#cb39-40"></a>PANGAEA_842237 <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_rds</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"_data_raw"</span>, <span class="st">"tara"</span>, <span class="st">"PANGAEA_842237.rds"</span>))</span>
<span id="cb39-41"><a href="#cb39-41"></a><span class="co"># carbonate chemistry</span></span>
<span id="cb39-42"><a href="#cb39-42"></a>PANGAEA_875567 <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_rds</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"_data_raw"</span>, <span class="st">"tara"</span>, <span class="st">"PANGAEA_875567.rds"</span>))</span>
<span id="cb39-43"><a href="#cb39-43"></a><span class="co"># nutrients</span></span>
<span id="cb39-44"><a href="#cb39-44"></a>PANGAEA_875575 <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_rds</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"_data_raw"</span>, <span class="st">"tara"</span>, <span class="st">"PANGAEA_875575.rds"</span>))</span>
<span id="cb39-45"><a href="#cb39-45"></a><span class="co"># phytoplankton pigments</span></span>
<span id="cb39-46"><a href="#cb39-46"></a>PANGAEA_875569 <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_rds</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"_data_raw"</span>, <span class="st">"tara"</span>, <span class="st">"PANGAEA_875569.rds"</span>))</span>
<span id="cb39-47"><a href="#cb39-47"></a><span class="co"># CTD sensor data</span></span>
<span id="cb39-48"><a href="#cb39-48"></a>PANGAEA_875576 <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_rds</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"_data_raw"</span>, <span class="st">"tara"</span>, <span class="st">"PANGAEA_875576.rds"</span>))</span>
<span id="cb39-49"><a href="#cb39-49"></a><span class="co"># mesoscale features</span></span>
<span id="cb39-50"><a href="#cb39-50"></a>PANGAEA_875577 <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_rds</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"_data_raw"</span>, <span class="st">"tara"</span>, <span class="st">"PANGAEA_875577.rds"</span>))</span>
<span id="cb39-51"><a href="#cb39-51"></a><span class="co"># whole water column features</span></span>
<span id="cb39-52"><a href="#cb39-52"></a>PANGAEA_875579 <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_rds</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"_data_raw"</span>, <span class="st">"tara"</span>, <span class="st">"PANGAEA_875579.rds"</span>))</span>
<span id="cb39-53"><a href="#cb39-53"></a><span class="in">```</span></span>
<span id="cb39-54"><a href="#cb39-54"></a></span>
<span id="cb39-55"><a href="#cb39-55"></a><span class="fu"># Collate, filter, and harmonize environmental data</span></span>
<span id="cb39-56"><a href="#cb39-56"></a></span>
<span id="cb39-57"><a href="#cb39-57"></a>In the final metagenomic tables we are interested in the following environmental variables:</span>
<span id="cb39-58"><a href="#cb39-58"></a></span>
<span id="cb39-59"><a href="#cb39-59"></a><span class="ss">1. </span>Density</span>
<span id="cb39-60"><a href="#cb39-60"></a><span class="ss">2. </span>Temperature</span>
<span id="cb39-61"><a href="#cb39-61"></a><span class="ss">3. </span>Salinity</span>
<span id="cb39-62"><a href="#cb39-62"></a><span class="ss">4. </span>Chlorophyll a</span>
<span id="cb39-63"><a href="#cb39-63"></a><span class="ss">5. </span>Dissolved oxygen (O2)</span>
<span id="cb39-64"><a href="#cb39-64"></a><span class="ss">6. </span>Total inorganic nitrogen (NO3 + NO2)</span>
<span id="cb39-65"><a href="#cb39-65"></a><span class="ss">7. </span>Nitrite (NO2)</span>
<span id="cb39-66"><a href="#cb39-66"></a><span class="ss">8. </span>Silica (SiO2)</span>
<span id="cb39-67"><a href="#cb39-67"></a><span class="ss">9. </span>Phosphate (PO4)</span>
<span id="cb39-68"><a href="#cb39-68"></a><span class="ss">10. </span>Total dissolved iron (FeT)</span>
<span id="cb39-69"><a href="#cb39-69"></a></span>
<span id="cb39-70"><a href="#cb39-70"></a>Thus we need to access the following environmental variables from various Pangaea dataframes for the final product</span>
<span id="cb39-71"><a href="#cb39-71"></a></span>
<span id="cb39-72"><a href="#cb39-72"></a>| Variable                                       | Dataset                            | Variable           |</span>
<span id="cb39-73"><a href="#cb39-73"></a>|------------------------------------------------|------------------------------------|--------------------|</span>
<span id="cb39-74"><a href="#cb39-74"></a>| Density                                        | Sensors, in situ (PANGAEA.875576)   | PANGAEA.875576_047 |</span>
<span id="cb39-75"><a href="#cb39-75"></a>| Salinity                                       | Sensors, in situ (PANGAEA.875576)   | PANGAEA.875576_037 |</span>
<span id="cb39-76"><a href="#cb39-76"></a>| Temperature                                    | Sensors, in situ (PANGAEA.875576)   | PANGAEA.875576_027 |</span>
<span id="cb39-77"><a href="#cb39-77"></a>| Dissolved Oxygen                               | Sensors, in situ (PANGAEA.875576)   | PANGAEA.875576_057 |</span>
<span id="cb39-78"><a href="#cb39-78"></a>| Chlorophyll a                                  | Sensors, in situ (PANGAEA.875576)   | PANGAEA.875576_077 |</span>
<span id="cb39-79"><a href="#cb39-79"></a>| Dissolved inorganic nitrogen (DIN = NO3 + NO2) | Nutrients, bottles (PANGAEA.875575) | PANGAEA.875575_031 |</span>
<span id="cb39-80"><a href="#cb39-80"></a>| Nitrite (NO2)                                  | Nutrients, bottles (PANGAEA.875575) | PANGAEA.875575_021 |</span>
<span id="cb39-81"><a href="#cb39-81"></a>| Phosphate (PO4)                                | Nutrients, bottles (PANGAEA.875575) | PANGAEA.875575_026 |</span>
<span id="cb39-82"><a href="#cb39-82"></a>| Silica (SiO2)                                  | Nutrients, bottles (PANGAEA.875575) | PANGAEA.875575_036 |</span>
<span id="cb39-83"><a href="#cb39-83"></a>| Darwin Model Total dissolved Iron (FeT)        | Modeled variables (PANGAEA.875577)  | PANGAEA.875577_039 |</span>
<span id="cb39-84"><a href="#cb39-84"></a>| Darwin Model NO2                               | Modeled variables (PANGAEA.875577)  | PANGAEA.875577_043 |</span>
<span id="cb39-85"><a href="#cb39-85"></a>| Darwin Model DIN                               | Modeled variables (PANGAEA.875577)  | PANGAEA.875577_045 |</span>
<span id="cb39-86"><a href="#cb39-86"></a>| Temp @ 10 meters                               | Water column (PANGAEA.875579)       | PANGAEA.875579_054 |</span>
<span id="cb39-87"><a href="#cb39-87"></a>| Temp @ mixed layer                             | Water column (PANGAEA.875579)       | PANGAEA.875579_057 |</span>
<span id="cb39-88"><a href="#cb39-88"></a>| Temp @ DCM                                     | Water column (PANGAEA.875579)       | PANGAEA.875579_059 |</span>
<span id="cb39-89"><a href="#cb39-89"></a>| Salinity @ 10 meters                           | Water column (PANGAEA.875579)       | PANGAEA.875579_064 |</span>
<span id="cb39-90"><a href="#cb39-90"></a>| Salinity @ mixed layer                         | Water column (PANGAEA.875579)       | PANGAEA.875579_067 |</span>
<span id="cb39-91"><a href="#cb39-91"></a>| Salinity @ DCM                                 | Water column (PANGAEA.875579)       | PANGAEA.875579_069 |</span>
<span id="cb39-92"><a href="#cb39-92"></a>| Density @ 10 meters                            | Water column (PANGAEA.875579)       | PANGAEA.875579_074 |</span>
<span id="cb39-93"><a href="#cb39-93"></a>| Density @ mixed layer                          | Water column (PANGAEA.875579)       | PANGAEA.875579_077 |</span>
<span id="cb39-94"><a href="#cb39-94"></a>| Density @ DCM                                  | Water column (PANGAEA.875579)       | PANGAEA.875579_079 |</span>
<span id="cb39-95"><a href="#cb39-95"></a>| Chlorophyll a @ 10 meters                      | Water column (PANGAEA.875579)       | PANGAEA.875579_094 |</span>
<span id="cb39-96"><a href="#cb39-96"></a>| Chlorophyll a @ mixed layer                    | Water column (PANGAEA.875579)       | PANGAEA.875579_097 |</span>
<span id="cb39-97"><a href="#cb39-97"></a>| Chlorophyll a @ DCM                            | Water column (PANGAEA.875579)       | PANGAEA.875579_099 |</span>
<span id="cb39-98"><a href="#cb39-98"></a>| Dissolved Oxygen @ 10 meters                   | Water column (PANGAEA.875579)       | PANGAEA.875579_104 |</span>
<span id="cb39-99"><a href="#cb39-99"></a>| Dissolved Oxygen @ mixed layer                 | Water column (PANGAEA.875579)       | PANGAEA.875579_107 |</span>
<span id="cb39-100"><a href="#cb39-100"></a>| Dissolved Oxygen @ DCM                         | Water column (PANGAEA.875579)       | PANGAEA.875579_109 |</span>
<span id="cb39-101"><a href="#cb39-101"></a>| DIN @ 10 meters                                | Water column (PANGAEA.875579)       | PANGAEA.875579_114 |</span>
<span id="cb39-102"><a href="#cb39-102"></a>| DIN @ mixed layer                              | Water column (PANGAEA.875579)       | PANGAEA.875579_117 |</span>
<span id="cb39-103"><a href="#cb39-103"></a>| DIN @ DCM                                      | Water column (PANGAEA.875579)       | PANGAEA.875579_119 |</span>
<span id="cb39-104"><a href="#cb39-104"></a></span>
<span id="cb39-105"><a href="#cb39-105"></a><span class="fu">## Combine all Pangaea datasets with BioSamples</span></span>
<span id="cb39-106"><a href="#cb39-106"></a></span>
<span id="cb39-109"><a href="#cb39-109"></a><span class="in">```{r}</span></span>
<span id="cb39-110"><a href="#cb39-110"></a>tara_biosamples_joined01 <span class="ot">&lt;-</span> <span class="fu">left_join</span>(tara_biosamples, PANGAEA_875575, <span class="at">by =</span> <span class="fu">join_by</span>(tara_barcode_num, tara_station)) <span class="sc">%&gt;%</span> </span>
<span id="cb39-111"><a href="#cb39-111"></a>  <span class="fu">left_join</span>(dplyr<span class="sc">::</span><span class="fu">select</span>(PANGAEA_875576, tara_barcode_num, tara_station, date_time, <span class="dv">21</span><span class="sc">:</span><span class="fu">last_col</span>()), </span>
<span id="cb39-112"><a href="#cb39-112"></a>            <span class="at">by =</span> <span class="fu">join_by</span>(tara_barcode_num, tara_station)) <span class="sc">%&gt;%</span> </span>
<span id="cb39-113"><a href="#cb39-113"></a>  <span class="fu">left_join</span>(dplyr<span class="sc">::</span><span class="fu">select</span>(PANGAEA_875577, tara_barcode_num, tara_station, <span class="dv">21</span><span class="sc">:</span><span class="fu">last_col</span>()), <span class="at">by =</span> <span class="fu">join_by</span>(tara_barcode_num, tara_station)) <span class="sc">%&gt;%</span> </span>
<span id="cb39-114"><a href="#cb39-114"></a>  <span class="fu">left_join</span>(dplyr<span class="sc">::</span><span class="fu">select</span>(PANGAEA_875579, tara_barcode_num, tara_station, <span class="dv">21</span><span class="sc">:</span><span class="fu">last_col</span>()), <span class="at">by =</span> <span class="fu">join_by</span>(tara_barcode_num, tara_station)) <span class="sc">%&gt;%</span> </span>
<span id="cb39-115"><a href="#cb39-115"></a>  <span class="fu">left_join</span>(dplyr<span class="sc">::</span><span class="fu">select</span>(PANGAEA_842237, tara_station, latitude, longitude), <span class="at">by =</span> <span class="fu">join_by</span>(tara_station))</span>
<span id="cb39-116"><a href="#cb39-116"></a><span class="in">```</span></span>
<span id="cb39-117"><a href="#cb39-117"></a></span>
<span id="cb39-118"><a href="#cb39-118"></a><span class="fu">## Fill missing observations with adjacent data</span></span>
<span id="cb39-119"><a href="#cb39-119"></a></span>
<span id="cb39-120"><a href="#cb39-120"></a>There are 7 cases where there is not single recorded depth for an observation, but rather a depth range (<span class="in">`depth_min`</span> to <span class="in">`depth_max`</span>).</span>
<span id="cb39-121"><a href="#cb39-121"></a></span>
<span id="cb39-124"><a href="#cb39-124"></a><span class="in">```{r}</span></span>
<span id="cb39-125"><a href="#cb39-125"></a>tara_biosamples_joined01 <span class="sc">%&gt;%</span> </span>
<span id="cb39-126"><a href="#cb39-126"></a>  <span class="fu">filter</span>(<span class="fu">is.na</span>(depth)) <span class="sc">%&gt;%</span> </span>
<span id="cb39-127"><a href="#cb39-127"></a>  <span class="fu">distinct</span>(tara_barcode_num, tara_station, env_ontology, depth, depth_min, depth_max, date_time)</span>
<span id="cb39-128"><a href="#cb39-128"></a><span class="in">```</span></span>
<span id="cb39-129"><a href="#cb39-129"></a></span>
<span id="cb39-130"><a href="#cb39-130"></a>Later it will make our lives easier if there are actual values for these depths and not NAs. We fill these depths with the mean of the depth range.</span>
<span id="cb39-131"><a href="#cb39-131"></a></span>
<span id="cb39-132"><a href="#cb39-132"></a>Also in some cases nutrients or sensor data will not be joined even though there is a measurement at that particular station/depth. We will group by station and depth and then fill those observations from the same station/depth. </span>
<span id="cb39-133"><a href="#cb39-133"></a></span>
<span id="cb39-136"><a href="#cb39-136"></a><span class="in">```{r}</span></span>
<span id="cb39-137"><a href="#cb39-137"></a>tara_biosamples_joined02 <span class="ot">&lt;-</span> tara_biosamples_joined01 <span class="sc">%&gt;%</span> </span>
<span id="cb39-138"><a href="#cb39-138"></a>  <span class="fu">mutate</span>(<span class="at">depth =</span> <span class="fu">if_else</span>(<span class="fu">is.na</span>(depth), (depth_max <span class="sc">-</span> depth_min)<span class="sc">/</span><span class="dv">2</span>, depth)) <span class="sc">%&gt;%</span> </span>
<span id="cb39-139"><a href="#cb39-139"></a>  <span class="fu">arrange</span>(tara_station, depth) <span class="sc">%&gt;%</span> </span>
<span id="cb39-140"><a href="#cb39-140"></a>  <span class="fu">group_by</span>(tara_station, depth, depth_min, depth_max) <span class="sc">%&gt;%</span> </span>
<span id="cb39-141"><a href="#cb39-141"></a>  <span class="fu">fill</span>(<span class="dv">20</span><span class="sc">:</span><span class="fu">last_col</span>(), <span class="at">.direction =</span> <span class="st">"downup"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb39-142"><a href="#cb39-142"></a>  <span class="fu">ungroup</span>()</span>
<span id="cb39-143"><a href="#cb39-143"></a><span class="in">```</span></span>
<span id="cb39-144"><a href="#cb39-144"></a></span>
<span id="cb39-145"><a href="#cb39-145"></a><span class="fu">## Fill missing observations with "whole water column" data (PANGAEA.875579)</span></span>
<span id="cb39-146"><a href="#cb39-146"></a></span>
<span id="cb39-147"><a href="#cb39-147"></a>Now we will fill missing observations with direct observations taken from the "whole water column."</span>
<span id="cb39-148"><a href="#cb39-148"></a></span>
<span id="cb39-149"><a href="#cb39-149"></a><span class="ss">1. </span>We will take bottle samples as the starting point (PANGAEA_875575)</span>
<span id="cb39-150"><a href="#cb39-150"></a><span class="ss">2. </span>If a bottle observation is missing fill with the closes CTD sensor observation (PANGAEA_875576)</span>
<span id="cb39-151"><a href="#cb39-151"></a><span class="ss">3. </span>If CTD sensor observation is missing fill with "whole water column" data (PANGAEA.875579). TBH I actually don't understand where this data comes from. The PI listed is Sabrina Speich who is different than the most common PIs (Pesant and Guidi). But from what I can tell this is not the output of a model but somehow a direct observation corresponding to a Tara station.</span>
<span id="cb39-152"><a href="#cb39-152"></a><span class="ss">4. </span>Finally, if none of the three above exist we will fill the data with the output from the MIT Darwin model.</span>
<span id="cb39-153"><a href="#cb39-153"></a></span>
<span id="cb39-154"><a href="#cb39-154"></a>This function creates a new variable replacing a variable from either PANGAEA_875575 or PANGAEA_875576 with the corresponding SRF/MIX/DCM/MES measurement from PANGAEA.875579</span>
<span id="cb39-155"><a href="#cb39-155"></a></span>
<span id="cb39-158"><a href="#cb39-158"></a><span class="in">```{r}</span></span>
<span id="cb39-159"><a href="#cb39-159"></a>replace_w_875579 <span class="ot">&lt;-</span> <span class="cf">function</span>(.data, varname, target, srf, mix, dcm){</span>
<span id="cb39-160"><a href="#cb39-160"></a>  .data <span class="sc">%&gt;%</span> </span>
<span id="cb39-161"><a href="#cb39-161"></a>    <span class="fu">mutate</span>(<span class="st">"{varname}"</span> <span class="sc">:</span><span class="er">=</span> <span class="fu">if_else</span>(<span class="sc">!</span><span class="fu">is.na</span>({{ target }}), {{ target }},</span>
<span id="cb39-162"><a href="#cb39-162"></a>          <span class="fu">if_else</span>(<span class="fu">str_detect</span>(env_ontology, <span class="st">"SRF"</span>), {{ srf }},</span>
<span id="cb39-163"><a href="#cb39-163"></a>                  <span class="fu">if_else</span>(<span class="fu">str_detect</span>(env_ontology, <span class="st">"MIX"</span>), {{ mix }}, </span>
<span id="cb39-164"><a href="#cb39-164"></a>                          <span class="fu">if_else</span>(<span class="fu">str_detect</span>(env_ontology, <span class="st">"DCM"</span>), {{ dcm }}, <span class="cn">NA_real_</span>)))))</span>
<span id="cb39-165"><a href="#cb39-165"></a>}</span>
<span id="cb39-166"><a href="#cb39-166"></a></span>
<span id="cb39-167"><a href="#cb39-167"></a>simple_replace <span class="ot">&lt;-</span> <span class="cf">function</span>(.data, varname, target, replace){</span>
<span id="cb39-168"><a href="#cb39-168"></a>  .data <span class="sc">%&gt;%</span> </span>
<span id="cb39-169"><a href="#cb39-169"></a>    <span class="fu">mutate</span>(<span class="st">"{varname}"</span> <span class="sc">:</span><span class="er">=</span> <span class="fu">if_else</span>(<span class="fu">is.na</span>({{ target }}), {{ replace }}, {{ target }}))</span>
<span id="cb39-170"><a href="#cb39-170"></a>}</span>
<span id="cb39-171"><a href="#cb39-171"></a><span class="in">```</span></span>
<span id="cb39-172"><a href="#cb39-172"></a></span>
<span id="cb39-173"><a href="#cb39-173"></a>Call the above function for each of the 10 variables of interest</span>
<span id="cb39-174"><a href="#cb39-174"></a></span>
<span id="cb39-177"><a href="#cb39-177"></a><span class="in">```{r}</span></span>
<span id="cb39-178"><a href="#cb39-178"></a>tara_biosamples_joined02_replaced <span class="ot">&lt;-</span> tara_biosamples_joined02 <span class="sc">%&gt;%</span></span>
<span id="cb39-179"><a href="#cb39-179"></a>  <span class="fu">replace_w_875579</span>(<span class="st">"density"</span>,     PANGAEA<span class="fl">.875576</span>_047, PANGAEA<span class="fl">.875579</span>_074, PANGAEA<span class="fl">.875579</span>_077, PANGAEA<span class="fl">.875579</span>_079) <span class="sc">%&gt;%</span> </span>
<span id="cb39-180"><a href="#cb39-180"></a>  <span class="fu">replace_w_875579</span>(<span class="st">"temperature"</span>, PANGAEA<span class="fl">.875576</span>_027, PANGAEA<span class="fl">.875579</span>_064, PANGAEA<span class="fl">.875579</span>_067, PANGAEA<span class="fl">.875579</span>_069) <span class="sc">%&gt;%</span> </span>
<span id="cb39-181"><a href="#cb39-181"></a>  <span class="fu">replace_w_875579</span>(<span class="st">"salinity"</span>,    PANGAEA<span class="fl">.875576</span>_037, PANGAEA<span class="fl">.875579</span>_054, PANGAEA<span class="fl">.875579</span>_057, PANGAEA<span class="fl">.875579</span>_059) <span class="sc">%&gt;%</span> </span>
<span id="cb39-182"><a href="#cb39-182"></a>  <span class="fu">replace_w_875579</span>(<span class="st">"chl_a"</span>,       PANGAEA<span class="fl">.875576</span>_077, PANGAEA<span class="fl">.875579</span>_094, PANGAEA<span class="fl">.875579</span>_097, PANGAEA<span class="fl">.875579</span>_099) <span class="sc">%&gt;%</span> </span>
<span id="cb39-183"><a href="#cb39-183"></a>  <span class="fu">replace_w_875579</span>(<span class="st">"dO2"</span>,         PANGAEA<span class="fl">.875576</span>_057, PANGAEA<span class="fl">.875579</span>_104, PANGAEA<span class="fl">.875579</span>_107, PANGAEA<span class="fl">.875579</span>_109) <span class="sc">%&gt;%</span> </span>
<span id="cb39-184"><a href="#cb39-184"></a>  <span class="fu">replace_w_875579</span>(<span class="st">"DIN"</span>,         PANGAEA<span class="fl">.875575</span>_031, PANGAEA<span class="fl">.875579</span>_114, PANGAEA<span class="fl">.875579</span>_117, PANGAEA<span class="fl">.875579</span>_119) <span class="sc">%&gt;%</span> </span>
<span id="cb39-185"><a href="#cb39-185"></a>    <span class="fu">simple_replace</span>(<span class="st">"DIN"</span>,         DIN, PANGAEA<span class="fl">.875577</span>_045) <span class="sc">%&gt;%</span> </span>
<span id="cb39-186"><a href="#cb39-186"></a>    <span class="fu">simple_replace</span>(<span class="st">"NO2"</span>,         PANGAEA<span class="fl">.875575</span>_021, PANGAEA<span class="fl">.875577</span>_043) <span class="sc">%&gt;%</span> </span>
<span id="cb39-187"><a href="#cb39-187"></a>  <span class="fu">mutate</span>(<span class="at">FeT =</span> PANGAEA<span class="fl">.875577</span>_039, </span>
<span id="cb39-188"><a href="#cb39-188"></a>         <span class="at">PO4 =</span> PANGAEA<span class="fl">.875575</span>_026,</span>
<span id="cb39-189"><a href="#cb39-189"></a>         <span class="at">SiO2 =</span> PANGAEA<span class="fl">.875575</span>_036) <span class="sc">%&gt;%</span> </span>
<span id="cb39-190"><a href="#cb39-190"></a>  <span class="fu">relocate</span>(tara_barcode_num, tara_station, env_ontology, latitude, longitude, date_time, depth, depth_min, depth_max, </span>
<span id="cb39-191"><a href="#cb39-191"></a>           density, temperature, salinity, chl_a, dO2, DIN, NO2, FeT, PO4, SiO2)</span>
<span id="cb39-192"><a href="#cb39-192"></a><span class="in">```</span></span>
<span id="cb39-193"><a href="#cb39-193"></a></span>
<span id="cb39-194"><a href="#cb39-194"></a>Fix some stuff with dates</span>
<span id="cb39-195"><a href="#cb39-195"></a></span>
<span id="cb39-198"><a href="#cb39-198"></a><span class="in">```{r}</span></span>
<span id="cb39-199"><a href="#cb39-199"></a>statistical_mode <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb39-200"><a href="#cb39-200"></a>  ux <span class="ot">&lt;-</span> <span class="fu">unique</span>(x)</span>
<span id="cb39-201"><a href="#cb39-201"></a>  ux[<span class="fu">which.max</span>(<span class="fu">tabulate</span>(<span class="fu">match</span>(x, ux)))]</span>
<span id="cb39-202"><a href="#cb39-202"></a>}</span>
<span id="cb39-203"><a href="#cb39-203"></a></span>
<span id="cb39-204"><a href="#cb39-204"></a>tara_biosamples_joined02_replaced <span class="ot">&lt;-</span> tara_biosamples_joined02_replaced <span class="sc">%&gt;%</span> </span>
<span id="cb39-205"><a href="#cb39-205"></a>  <span class="co"># convert the date_time to only date. This ensures that when you are grouping</span></span>
<span id="cb39-206"><a href="#cb39-206"></a>  <span class="co"># by tara_station and date_time that you don't expand the results to include</span></span>
<span id="cb39-207"><a href="#cb39-207"></a>  <span class="co"># multiple times within a single station</span></span>
<span id="cb39-208"><a href="#cb39-208"></a>  <span class="fu">mutate</span>(<span class="at">date_time =</span> lubridate<span class="sc">::</span><span class="fu">as_date</span>(stringr<span class="sc">::</span><span class="fu">str_extract</span>(date_time, <span class="st">"^</span><span class="sc">\\</span><span class="st">d{4}-</span><span class="sc">\\</span><span class="st">d{2}-</span><span class="sc">\\</span><span class="st">d{2}"</span>))) <span class="sc">%&gt;%</span> </span>
<span id="cb39-209"><a href="#cb39-209"></a>  <span class="fu">group_by</span>(tara_station) <span class="sc">%&gt;%</span> </span>
<span id="cb39-210"><a href="#cb39-210"></a>  <span class="co"># in the case that there are multiple days sampled at a station take the most</span></span>
<span id="cb39-211"><a href="#cb39-211"></a>  <span class="co"># common day to represent the whole station</span></span>
<span id="cb39-212"><a href="#cb39-212"></a>  <span class="fu">mutate</span>(<span class="at">date_time =</span> <span class="fu">statistical_mode</span>(date_time)) <span class="sc">%&gt;%</span> </span>
<span id="cb39-213"><a href="#cb39-213"></a>  <span class="fu">ungroup</span>()</span>
<span id="cb39-214"><a href="#cb39-214"></a><span class="in">```</span></span>
<span id="cb39-215"><a href="#cb39-215"></a></span>
<span id="cb39-216"><a href="#cb39-216"></a>Now after we have filled observations with data internally available from the Tara Pangaea collection, we need to find which stations still have missing data observations for one or more these variables.</span>
<span id="cb39-217"><a href="#cb39-217"></a></span>
<span id="cb39-220"><a href="#cb39-220"></a><span class="in">```{r}</span></span>
<span id="cb39-221"><a href="#cb39-221"></a>target_vars <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"density"</span>, <span class="st">"temperature"</span>, <span class="st">"salinity"</span>, <span class="st">"chl_a"</span>, <span class="st">"dO2"</span>, <span class="st">"DIN"</span>, <span class="st">"NO2"</span>, <span class="st">"FeT"</span>, <span class="st">"PO4"</span>, <span class="st">"SiO2"</span>)</span>
<span id="cb39-222"><a href="#cb39-222"></a></span>
<span id="cb39-223"><a href="#cb39-223"></a>missing <span class="ot">&lt;-</span> tara_biosamples_joined02_replaced <span class="sc">%&gt;%</span> </span>
<span id="cb39-224"><a href="#cb39-224"></a>  <span class="fu">filter</span>(<span class="fu">if_any</span>(<span class="fu">all_of</span>(target_vars), is.na)) <span class="sc">%&gt;%</span></span>
<span id="cb39-225"><a href="#cb39-225"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="fu">all_of</span>(target_vars), <span class="sc">~</span><span class="fu">any</span>(<span class="fu">is.na</span>(.))),</span>
<span id="cb39-226"><a href="#cb39-226"></a>            <span class="at">.by =</span> <span class="fu">c</span>(tara_station, env_ontology, depth, depth_min, depth_max, latitude, longitude, date_time))</span>
<span id="cb39-227"><a href="#cb39-227"></a><span class="in">```</span></span>
<span id="cb39-228"><a href="#cb39-228"></a></span>
<span id="cb39-229"><a href="#cb39-229"></a>Most missing observations are because of PO4 and SiO2</span>
<span id="cb39-230"><a href="#cb39-230"></a></span>
<span id="cb39-233"><a href="#cb39-233"></a><span class="in">```{r}</span></span>
<span id="cb39-234"><a href="#cb39-234"></a>missing <span class="sc">%&gt;%</span> </span>
<span id="cb39-235"><a href="#cb39-235"></a>  <span class="fu">summarize</span>(<span class="fu">across</span>(<span class="fu">all_of</span>(target_vars), sum))</span>
<span id="cb39-236"><a href="#cb39-236"></a><span class="in">```</span></span>
<span id="cb39-237"><a href="#cb39-237"></a></span>
<span id="cb39-238"><a href="#cb39-238"></a>And this is for 56 different tara stations</span>
<span id="cb39-239"><a href="#cb39-239"></a></span>
<span id="cb39-242"><a href="#cb39-242"></a><span class="in">```{r}</span></span>
<span id="cb39-243"><a href="#cb39-243"></a><span class="fu">distinct</span>(missing, tara_station) <span class="sc">%&gt;%</span> </span>
<span id="cb39-244"><a href="#cb39-244"></a>  <span class="fu">count</span>()</span>
<span id="cb39-245"><a href="#cb39-245"></a><span class="in">```</span></span>
<span id="cb39-246"><a href="#cb39-246"></a></span>
<span id="cb39-247"><a href="#cb39-247"></a>This level of completion is pretty OK I think... To replace the missing fields I will use modeled variables from the MIT Darwin model. These can be obtained from <span class="co">[</span><span class="ot">Simons CMAP.</span><span class="co">](https://simonscmap.com/)</span></span>
<span id="cb39-248"><a href="#cb39-248"></a></span>
<span id="cb39-249"><a href="#cb39-249"></a>Set the api-key for CMAP. This allows your R session to communicate with their backend database. You will need to get your own key following instructions <span class="co">[</span><span class="ot">here</span><span class="co">](https://simonscmap.com/apikeymanagement)</span>.</span>
<span id="cb39-250"><a href="#cb39-250"></a></span>
<span id="cb39-253"><a href="#cb39-253"></a><span class="in">```{r}</span></span>
<span id="cb39-254"><a href="#cb39-254"></a><span class="co">#| echo: true</span></span>
<span id="cb39-255"><a href="#cb39-255"></a><span class="co">#| eval: false</span></span>
<span id="cb39-256"><a href="#cb39-256"></a><span class="co">#| output: false</span></span>
<span id="cb39-257"><a href="#cb39-257"></a><span class="fu">library</span>(cmap4r)</span>
<span id="cb39-258"><a href="#cb39-258"></a>cmap4r<span class="sc">::</span><span class="fu">set_authorization</span>(<span class="at">reset =</span> <span class="cn">TRUE</span>)</span>
<span id="cb39-259"><a href="#cb39-259"></a>cmap4r<span class="sc">::</span><span class="fu">set_authorization</span>(<span class="at">cmap_key =</span> <span class="st">"KEY-GOES-HERE"</span>)</span>
<span id="cb39-260"><a href="#cb39-260"></a><span class="in">```</span></span>
<span id="cb39-261"><a href="#cb39-261"></a></span>
<span id="cb39-262"><a href="#cb39-262"></a>Download a local copy of the CMAP catalog. Useful for querying later...</span>
<span id="cb39-263"><a href="#cb39-263"></a></span>
<span id="cb39-266"><a href="#cb39-266"></a><span class="in">```{r}</span></span>
<span id="cb39-267"><a href="#cb39-267"></a><span class="co">#| echo: true</span></span>
<span id="cb39-268"><a href="#cb39-268"></a><span class="co">#| eval: false</span></span>
<span id="cb39-269"><a href="#cb39-269"></a><span class="co">#| output: false</span></span>
<span id="cb39-270"><a href="#cb39-270"></a>local_cat <span class="ot">&lt;-</span> cmap4r<span class="sc">::</span><span class="fu">get_catalog</span>() <span class="sc">%&gt;%</span></span>
<span id="cb39-271"><a href="#cb39-271"></a>  <span class="fu">select</span>(Variable, Table_Name, Unit, Sensor, Unit)</span>
<span id="cb39-272"><a href="#cb39-272"></a></span>
<span id="cb39-273"><a href="#cb39-273"></a><span class="fu">write_tsv</span>(local_cat, here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"_data_raw"</span>, <span class="st">"cmap"</span>, <span class="st">"cmap_catalog.tsv"</span>))</span>
<span id="cb39-274"><a href="#cb39-274"></a><span class="in">```</span></span>
<span id="cb39-275"><a href="#cb39-275"></a></span>
<span id="cb39-278"><a href="#cb39-278"></a><span class="in">```{r}</span></span>
<span id="cb39-279"><a href="#cb39-279"></a><span class="co">#| echo: false</span></span>
<span id="cb39-280"><a href="#cb39-280"></a><span class="co">#| eval: true</span></span>
<span id="cb39-281"><a href="#cb39-281"></a><span class="co">#| output: false</span></span>
<span id="cb39-282"><a href="#cb39-282"></a>local_cat <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"_data_raw"</span>, <span class="st">"cmap"</span>, <span class="st">"cmap_catalog.tsv"</span>))</span>
<span id="cb39-283"><a href="#cb39-283"></a><span class="in">```</span></span>
<span id="cb39-284"><a href="#cb39-284"></a></span>
<span id="cb39-285"><a href="#cb39-285"></a><span class="fu">## Fill missing observations with Darwin model Nutrients output</span></span>
<span id="cb39-286"><a href="#cb39-286"></a></span>
<span id="cb39-287"><a href="#cb39-287"></a><span class="fu">### Darwin coordinates for nutrients</span></span>
<span id="cb39-288"><a href="#cb39-288"></a></span>
<span id="cb39-289"><a href="#cb39-289"></a>CMAP has some select nutrient concentrations available from the <span class="co">[</span><span class="ot">3-day averagedDarwin Model output</span><span class="co">](https://simonscmap.com/catalog/datasets/Darwin_Nutrient)</span>. These include</span>
<span id="cb39-290"><a href="#cb39-290"></a></span>
<span id="cb39-293"><a href="#cb39-293"></a><span class="in">```{r}</span></span>
<span id="cb39-294"><a href="#cb39-294"></a>local_cat <span class="sc">%&gt;%</span> </span>
<span id="cb39-295"><a href="#cb39-295"></a>  <span class="fu">filter</span>(Table_Name <span class="sc">==</span> <span class="st">"tblDarwin_Nutrient"</span>)</span>
<span id="cb39-296"><a href="#cb39-296"></a><span class="in">```</span></span>
<span id="cb39-297"><a href="#cb39-297"></a></span>
<span id="cb39-298"><a href="#cb39-298"></a>Not exactly clear what the units are for these nutrients, but my guess is mmol m-3 since most other concentration units are per 1 m3 unit volume.</span>
<span id="cb39-299"><a href="#cb39-299"></a></span>
<span id="cb39-300"><a href="#cb39-300"></a>Now we will get only the stations with missing nutrients in <span class="in">`tblDarwin_Nutrient`</span></span>
<span id="cb39-301"><a href="#cb39-301"></a></span>
<span id="cb39-304"><a href="#cb39-304"></a><span class="in">```{r}</span></span>
<span id="cb39-305"><a href="#cb39-305"></a>missing_nuts <span class="ot">&lt;-</span> missing <span class="sc">%&gt;%</span> </span>
<span id="cb39-306"><a href="#cb39-306"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="fu">c</span>(<span class="sc">-</span>tara_station, <span class="sc">-</span>env_ontology, <span class="sc">-</span>depth, <span class="sc">-</span>depth_min, <span class="sc">-</span>depth_max, <span class="sc">-</span>latitude, <span class="sc">-</span>longitude, <span class="sc">-</span>date_time)) <span class="sc">%&gt;%</span> </span>
<span id="cb39-307"><a href="#cb39-307"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(value <span class="sc">==</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb39-308"><a href="#cb39-308"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(tara_station, env_ontology, latitude, longitude, date_time, depth, depth_min, depth_max) <span class="sc">%&gt;%</span> </span>
<span id="cb39-309"><a href="#cb39-309"></a>  dplyr<span class="sc">::</span><span class="fu">distinct</span>()</span>
<span id="cb39-310"><a href="#cb39-310"></a><span class="in">```</span></span>
<span id="cb39-311"><a href="#cb39-311"></a></span>
<span id="cb39-312"><a href="#cb39-312"></a>Now retrieve parameters from these stations that can be used to query CMAP database. We need a start latitude, ending latitude, starting longitude, ending longitude, starting depth, ending depth, and starting date and ending date. Note: here we don't bother with depths because when you query all the variables in the CMAP database it returns all depths regardless of what you provide to the function (bug? feature?)</span>
<span id="cb39-313"><a href="#cb39-313"></a></span>
<span id="cb39-316"><a href="#cb39-316"></a><span class="in">```{r}</span></span>
<span id="cb39-317"><a href="#cb39-317"></a>coord_range <span class="ot">&lt;-</span> <span class="cf">function</span>(x, f, offset, <span class="at">negate =</span> <span class="cn">FALSE</span>){</span>
<span id="cb39-318"><a href="#cb39-318"></a>  <span class="fu">return</span>(<span class="fu">if_else</span>(negate, <span class="sc">-</span>(<span class="fu">f</span>(<span class="fu">max</span>(<span class="fu">abs</span>(x))) <span class="sc">+</span> offset), <span class="fu">f</span>(<span class="fu">max</span>(<span class="fu">abs</span>(x))) <span class="sc">+</span> offset))</span>
<span id="cb39-319"><a href="#cb39-319"></a>}</span>
<span id="cb39-320"><a href="#cb39-320"></a></span>
<span id="cb39-321"><a href="#cb39-321"></a>missing_nuts_nested <span class="ot">&lt;-</span> missing_nuts <span class="sc">%&gt;%</span> </span>
<span id="cb39-322"><a href="#cb39-322"></a>  <span class="fu">group_by</span>(tara_station) <span class="sc">%&gt;%</span> </span>
<span id="cb39-323"><a href="#cb39-323"></a>  <span class="fu">mutate</span>(<span class="at">lat1 =</span> <span class="fu">if_else</span>(latitude <span class="sc">&lt;</span> <span class="dv">0</span>,   <span class="fu">coord_range</span>(latitude, ceiling, <span class="dv">1</span>, <span class="cn">TRUE</span>),  <span class="fu">coord_range</span>(latitude, floor, <span class="sc">-</span><span class="dv">1</span>, <span class="cn">FALSE</span>)),</span>
<span id="cb39-324"><a href="#cb39-324"></a>         <span class="at">lat2 =</span> <span class="fu">if_else</span>(latitude <span class="sc">&lt;</span> <span class="dv">0</span>,   <span class="fu">coord_range</span>(latitude, floor, <span class="sc">-</span><span class="dv">1</span>, <span class="cn">TRUE</span>),   <span class="fu">coord_range</span>(latitude, ceiling, <span class="dv">1</span>, <span class="cn">FALSE</span>)),</span>
<span id="cb39-325"><a href="#cb39-325"></a>         <span class="at">lon1 =</span> <span class="fu">if_else</span>(longitude <span class="sc">&lt;</span> <span class="dv">0</span>,  <span class="fu">coord_range</span>(longitude, ceiling, <span class="dv">1</span>, <span class="cn">TRUE</span>),  <span class="fu">coord_range</span>(longitude, floor, <span class="sc">-</span><span class="dv">1</span>, <span class="cn">FALSE</span>)),</span>
<span id="cb39-326"><a href="#cb39-326"></a>         <span class="at">lon2 =</span> <span class="fu">if_else</span>(longitude <span class="sc">&lt;</span> <span class="dv">0</span>,  <span class="fu">coord_range</span>(longitude, floor, <span class="sc">-</span><span class="dv">1</span>, <span class="cn">TRUE</span>), <span class="fu">coord_range</span>(longitude, ceiling, <span class="dv">1</span>, <span class="cn">FALSE</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb39-327"><a href="#cb39-327"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">dt1 =</span> <span class="fu">floor_date</span>(date_time, <span class="st">'month'</span>),</span>
<span id="cb39-328"><a href="#cb39-328"></a>                <span class="at">dt2 =</span> <span class="fu">ceiling_date</span>(date_time, <span class="st">'month'</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb39-329"><a href="#cb39-329"></a>  <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb39-330"><a href="#cb39-330"></a>  <span class="fu">nest</span>(<span class="at">original_data =</span> <span class="fu">c</span>(env_ontology<span class="sc">:</span>depth_max)) <span class="sc">%&gt;%</span> </span>
<span id="cb39-331"><a href="#cb39-331"></a>  <span class="fu">nest</span>(<span class="at">dw_query =</span> <span class="fu">c</span>(lat1<span class="sc">:</span>dt2))</span>
<span id="cb39-332"><a href="#cb39-332"></a><span class="in">```</span></span>
<span id="cb39-333"><a href="#cb39-333"></a></span>
<span id="cb39-334"><a href="#cb39-334"></a></span>
<span id="cb39-335"><a href="#cb39-335"></a>Now make the call querying the CMAP remote database. This maps a separate call to every unique missing Tara Station. Even though this isn't very efficient it is faster and takes less space that just downloaded the Darwin output for the entire ocean and then later filtering to the coordinates of interest.</span>
<span id="cb39-336"><a href="#cb39-336"></a></span>
<span id="cb39-339"><a href="#cb39-339"></a><span class="in">```{r}</span></span>
<span id="cb39-340"><a href="#cb39-340"></a><span class="co">#| echo: true</span></span>
<span id="cb39-341"><a href="#cb39-341"></a><span class="co">#| eval: false</span></span>
<span id="cb39-342"><a href="#cb39-342"></a><span class="co">#| output: false</span></span>
<span id="cb39-343"><a href="#cb39-343"></a>missing_nuts_nested_darwin <span class="ot">&lt;-</span> missing_nuts_nested <span class="sc">%&gt;%</span> </span>
<span id="cb39-344"><a href="#cb39-344"></a>  <span class="fu">mutate</span>(</span>
<span id="cb39-345"><a href="#cb39-345"></a>    <span class="at">tbl =</span> <span class="fu">map</span>(</span>
<span id="cb39-346"><a href="#cb39-346"></a>      dw_query,</span>
<span id="cb39-347"><a href="#cb39-347"></a>      <span class="cf">function</span>(df) cmap4r<span class="sc">::</span><span class="fu">get_spacetime</span>(</span>
<span id="cb39-348"><a href="#cb39-348"></a>        <span class="at">tableName =</span> <span class="st">'tblDarwin_Nutrient'</span>,</span>
<span id="cb39-349"><a href="#cb39-349"></a>        <span class="at">varName =</span> <span class="st">'*'</span>,</span>
<span id="cb39-350"><a href="#cb39-350"></a>        <span class="at">dt1 =</span> <span class="fu">as.character</span>(df<span class="sc">$</span>dt1),</span>
<span id="cb39-351"><a href="#cb39-351"></a>        <span class="at">dt2 =</span> <span class="fu">as.character</span>(df<span class="sc">$</span>dt2),</span>
<span id="cb39-352"><a href="#cb39-352"></a>        <span class="at">lat1 =</span> df<span class="sc">$</span>lat1,</span>
<span id="cb39-353"><a href="#cb39-353"></a>        <span class="at">lat2 =</span> df<span class="sc">$</span>lat2,</span>
<span id="cb39-354"><a href="#cb39-354"></a>        <span class="at">lon1 =</span> df<span class="sc">$</span>lon1,</span>
<span id="cb39-355"><a href="#cb39-355"></a>        <span class="at">lon2 =</span> df<span class="sc">$</span>lon2)))</span>
<span id="cb39-356"><a href="#cb39-356"></a></span>
<span id="cb39-357"><a href="#cb39-357"></a><span class="co"># save this for later to speed up quarto renders</span></span>
<span id="cb39-358"><a href="#cb39-358"></a><span class="fu">write_rds</span>(missing_nuts_nested_darwin, here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"_data_raw"</span>, <span class="st">"tara"</span>, <span class="st">"cmap_missing_nuts.rds"</span>))</span>
<span id="cb39-359"><a href="#cb39-359"></a><span class="in">```</span></span>
<span id="cb39-360"><a href="#cb39-360"></a></span>
<span id="cb39-363"><a href="#cb39-363"></a><span class="in">```{r}</span></span>
<span id="cb39-364"><a href="#cb39-364"></a><span class="co">#| echo: false</span></span>
<span id="cb39-365"><a href="#cb39-365"></a><span class="co">#| eval: true</span></span>
<span id="cb39-366"><a href="#cb39-366"></a><span class="co">#| output: false</span></span>
<span id="cb39-367"><a href="#cb39-367"></a>missing_nuts_nested_darwin <span class="ot">&lt;-</span> <span class="fu">read_rds</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"_data_raw"</span>, <span class="st">"tara"</span>, <span class="st">"cmap_missing_nuts.rds"</span>))</span>
<span id="cb39-368"><a href="#cb39-368"></a><span class="in">```</span></span>
<span id="cb39-369"><a href="#cb39-369"></a></span>
<span id="cb39-370"><a href="#cb39-370"></a><span class="fu">### Filter Darwin output to Tara coordinates</span></span>
<span id="cb39-371"><a href="#cb39-371"></a></span>
<span id="cb39-372"><a href="#cb39-372"></a>Now we need to whittle down the Darwin model output to only the time, depth, and coordinates of interest. We do this by </span>
<span id="cb39-373"><a href="#cb39-373"></a></span>
<span id="cb39-374"><a href="#cb39-374"></a><span class="ss">1. </span>Fuzzy joining on latitude and longitude using <span class="in">`fuzzyjoin::geo_inner_join`</span>. Briefly this calculates the <span class="co">[</span><span class="ot">Haversine distance</span><span class="co">](https://en.wikipedia.org/wiki/Haversine_formula)</span> between two sets of coordinates and keeps only those with distance magnitude less than <span class="in">`max_dist`</span> (here we set this to 65 kilometers). This allows us to keep only coordinates that are as close as possible to the sampled Tara coordinates. Later we select for the smallest Haversine distance to select a point from the model grid that is closest to the Tara station. Sometimes these coordinates will be a land cell on the Darwin grid and will return NAs so we use <span class="in">`drop_na`</span> to only include coordinates in the ocean with nutrients.</span>
<span id="cb39-375"><a href="#cb39-375"></a><span class="ss">2. </span>Next we calculate the duration of time between the Tara observation and the time step from the Darwin Model. Note this is the 3-day averaged model output. We then filter to only include model output that is closest in time (shortest duration) to the Tara observation.</span>
<span id="cb39-376"><a href="#cb39-376"></a><span class="ss">3. </span>Finally, we need to make a measure of the distance between the sampled Tara depth and the model output depth. In the upper 250 the depth is high resolution so usually we find a depth between <span class="in">`depth_min`</span> and <span class="in">`depth_max`</span> in the model output. Deeper in the ocean the model output is only every 100 meters or more so we need to filter on the smallest possible difference from the Tara depth</span>
<span id="cb39-377"><a href="#cb39-377"></a></span>
<span id="cb39-380"><a href="#cb39-380"></a><span class="in">```{r}</span></span>
<span id="cb39-381"><a href="#cb39-381"></a>missing_nuts_nested_darwin_filtered <span class="ot">&lt;-</span> missing_nuts_nested_darwin <span class="sc">%&gt;%</span> </span>
<span id="cb39-382"><a href="#cb39-382"></a>  <span class="co"># unnest on the Darwin output. expand rows per Tara Station</span></span>
<span id="cb39-383"><a href="#cb39-383"></a>  <span class="fu">unnest</span>(tbl) <span class="sc">%&gt;%</span></span>
<span id="cb39-384"><a href="#cb39-384"></a>  <span class="co"># select only relevant columns to make management easier</span></span>
<span id="cb39-385"><a href="#cb39-385"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(tara_station, </span>
<span id="cb39-386"><a href="#cb39-386"></a>                <span class="at">date_time =</span> time, <span class="at">latitude =</span> lat, <span class="at">longitude =</span> lon, depth,</span>
<span id="cb39-387"><a href="#cb39-387"></a>                <span class="at">PO4_dw =</span> PO4, <span class="at">SiO2_dw =</span> SiO2, <span class="at">O2_dw =</span> O2) <span class="sc">%&gt;%</span></span>
<span id="cb39-388"><a href="#cb39-388"></a>  <span class="co"># fuzzy join the Darwin output to the coordinates of missing data using the </span></span>
<span id="cb39-389"><a href="#cb39-389"></a>  <span class="co"># haversine distance and excluding joins that exceed 65 km in distance apart</span></span>
<span id="cb39-390"><a href="#cb39-390"></a>  fuzzyjoin<span class="sc">::</span><span class="fu">geo_inner_join</span>(missing_nuts,</span>
<span id="cb39-391"><a href="#cb39-391"></a>                            <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"latitude"</span>, <span class="st">"longitude"</span>),</span>
<span id="cb39-392"><a href="#cb39-392"></a>                            <span class="at">distance_col =</span> <span class="st">"geojoin_dist_km"</span>,</span>
<span id="cb39-393"><a href="#cb39-393"></a>                            <span class="at">max_dist =</span> <span class="dv">65</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb39-394"><a href="#cb39-394"></a>  <span class="co"># narrow down the data so we only consider matching Tara stations from the two datasets</span></span>
<span id="cb39-395"><a href="#cb39-395"></a>  <span class="fu">filter</span>(tara_station.x <span class="sc">==</span> tara_station.y) <span class="sc">%&gt;%</span></span>
<span id="cb39-396"><a href="#cb39-396"></a>  <span class="co"># exclude any Darwin observations that have NAs for PO4, SiO2, or O2 (these are genarally land grids or seafloor)</span></span>
<span id="cb39-397"><a href="#cb39-397"></a>  <span class="fu">filter</span>(<span class="sc">!</span>(<span class="fu">is.na</span>(PO4_dw) <span class="sc">&amp;</span> <span class="fu">is.na</span>(SiO2_dw) <span class="sc">&amp;</span> <span class="fu">is.na</span>(O2_dw))) <span class="sc">%&gt;%</span> </span>
<span id="cb39-398"><a href="#cb39-398"></a>  <span class="co"># Group by unique variables for each Tara observation</span></span>
<span id="cb39-399"><a href="#cb39-399"></a>  <span class="fu">group_by</span>(tara_station.y, env_ontology, depth.y, depth_min, depth_max) <span class="sc">%&gt;%</span> </span>
<span id="cb39-400"><a href="#cb39-400"></a>  <span class="co"># Filter to include only matches with smallest distance between Darwin grid and the coords of the direct observation</span></span>
<span id="cb39-401"><a href="#cb39-401"></a>  <span class="fu">filter</span>(geojoin_dist_km <span class="sc">==</span> <span class="fu">min</span>(geojoin_dist_km)) <span class="sc">%&gt;%</span> </span>
<span id="cb39-402"><a href="#cb39-402"></a>  <span class="co"># do some type conversions for dates to allow subsequent calculations</span></span>
<span id="cb39-403"><a href="#cb39-403"></a>  <span class="fu">mutate</span>(<span class="at">date_time.x =</span> lubridate<span class="sc">::</span><span class="fu">as_date</span>(date_time.x),</span>
<span id="cb39-404"><a href="#cb39-404"></a>         <span class="at">date_time.y =</span> lubridate<span class="sc">::</span><span class="fu">as_date</span>(stringr<span class="sc">::</span><span class="fu">str_extract</span>(date_time.y, <span class="st">"^</span><span class="sc">\\</span><span class="st">d{4}-</span><span class="sc">\\</span><span class="st">d{2}-</span><span class="sc">\\</span><span class="st">d{2}"</span>))) <span class="sc">%&gt;%</span> </span>
<span id="cb39-405"><a href="#cb39-405"></a>  <span class="co"># calculate the difference between the model output (3-day interval) and the sampling time of the direct observation</span></span>
<span id="cb39-406"><a href="#cb39-406"></a>  <span class="fu">mutate</span>(<span class="at">time_diff =</span> <span class="fu">as.duration</span>(<span class="fu">abs</span>(<span class="fu">interval</span>(date_time.x, date_time.y)))) <span class="sc">%&gt;%</span> </span>
<span id="cb39-407"><a href="#cb39-407"></a>  <span class="co"># filter to only include matches with the smallest time difference</span></span>
<span id="cb39-408"><a href="#cb39-408"></a>  <span class="fu">filter</span>(time_diff <span class="sc">==</span> <span class="fu">min</span>(time_diff)) <span class="sc">%&gt;%</span> </span>
<span id="cb39-409"><a href="#cb39-409"></a>  <span class="co"># filter to only include matches with the smallest depth difference</span></span>
<span id="cb39-410"><a href="#cb39-410"></a>  <span class="fu">filter</span>(<span class="fu">abs</span>(depth.y <span class="sc">-</span> depth.x) <span class="sc">==</span> <span class="fu">min</span>(<span class="fu">abs</span>(depth.y <span class="sc">-</span> depth.x))) <span class="sc">%&gt;%</span> </span>
<span id="cb39-411"><a href="#cb39-411"></a>  <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb39-412"><a href="#cb39-412"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="at">tara_station =</span> tara_station.y, </span>
<span id="cb39-413"><a href="#cb39-413"></a>                <span class="at">depth =</span> depth.y, depth_min, depth_max, env_ontology,</span>
<span id="cb39-414"><a href="#cb39-414"></a>                <span class="at">latitude =</span> latitude.y, <span class="at">longitude =</span> longitude.y, PO4_dw, SiO2_dw, O2_dw) <span class="sc">%&gt;%</span></span>
<span id="cb39-415"><a href="#cb39-415"></a>  <span class="co"># take mean of multi-observations (e.g. if the real depth is 4, the closest</span></span>
<span id="cb39-416"><a href="#cb39-416"></a>  <span class="co"># depths from Darwin are 35 and 45 so we average the concentration from the</span></span>
<span id="cb39-417"><a href="#cb39-417"></a>  <span class="co"># two flanking depths)</span></span>
<span id="cb39-418"><a href="#cb39-418"></a>  <span class="fu">summarize</span>(<span class="fu">across</span>(PO4_dw<span class="sc">:</span>O2_dw, mean), </span>
<span id="cb39-419"><a href="#cb39-419"></a>            <span class="at">.by =</span> <span class="fu">c</span>(tara_station<span class="sc">:</span>longitude))</span>
<span id="cb39-420"><a href="#cb39-420"></a><span class="in">```</span></span>
<span id="cb39-421"><a href="#cb39-421"></a></span>
<span id="cb39-422"><a href="#cb39-422"></a><span class="fu">### Join observed and Darwin Model output into single dataframe</span></span>
<span id="cb39-423"><a href="#cb39-423"></a></span>
<span id="cb39-426"><a href="#cb39-426"></a><span class="in">```{r}</span></span>
<span id="cb39-427"><a href="#cb39-427"></a>tara_biosamples_joined02_replaced02 <span class="ot">&lt;-</span> <span class="fu">left_join</span>(tara_biosamples_joined02_replaced, missing_nuts_nested_darwin_filtered, </span>
<span id="cb39-428"><a href="#cb39-428"></a>          <span class="at">by =</span> <span class="fu">join_by</span>(tara_station, env_ontology, latitude, longitude, depth, depth_min, depth_max)) <span class="sc">%&gt;%</span> </span>
<span id="cb39-429"><a href="#cb39-429"></a>  <span class="fu">simple_replace</span>(<span class="st">"SiO2"</span>, SiO2, SiO2_dw) <span class="sc">%&gt;%</span></span>
<span id="cb39-430"><a href="#cb39-430"></a>  <span class="fu">simple_replace</span>(<span class="st">"dO2"</span>,  dO2, O2_dw) <span class="sc">%&gt;%</span></span>
<span id="cb39-431"><a href="#cb39-431"></a>  <span class="fu">simple_replace</span>(<span class="st">"PO4"</span>,  PO4, PO4_dw)</span>
<span id="cb39-432"><a href="#cb39-432"></a><span class="in">```</span></span>
<span id="cb39-433"><a href="#cb39-433"></a></span>
<span id="cb39-434"><a href="#cb39-434"></a><span class="fu">## Impute remaining missing observations</span></span>
<span id="cb39-435"><a href="#cb39-435"></a></span>
<span id="cb39-438"><a href="#cb39-438"></a><span class="in">```{r}</span></span>
<span id="cb39-439"><a href="#cb39-439"></a><span class="fu">library</span>(missRanger)</span>
<span id="cb39-440"><a href="#cb39-440"></a><span class="in">```</span></span>
<span id="cb39-441"><a href="#cb39-441"></a></span>
<span id="cb39-442"><a href="#cb39-442"></a>There are three Tara stations with missing observations. <span class="in">`TARA_011`</span> is missing chl_a, density, salinity and temperature, while the others are missing only chl_a</span>
<span id="cb39-443"><a href="#cb39-443"></a></span>
<span id="cb39-446"><a href="#cb39-446"></a><span class="in">```{r}</span></span>
<span id="cb39-447"><a href="#cb39-447"></a>missing <span class="sc">%&gt;%</span> </span>
<span id="cb39-448"><a href="#cb39-448"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(tara_station<span class="sc">:</span>chl_a) <span class="sc">%&gt;%</span></span>
<span id="cb39-449"><a href="#cb39-449"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="fu">c</span>(<span class="sc">-</span>tara_station, <span class="sc">-</span>env_ontology, <span class="sc">-</span>depth, <span class="sc">-</span>depth_min, <span class="sc">-</span>depth_max, <span class="sc">-</span>latitude, <span class="sc">-</span>longitude, <span class="sc">-</span>date_time)) <span class="sc">%&gt;%</span> </span>
<span id="cb39-450"><a href="#cb39-450"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(value <span class="sc">==</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb39-451"><a href="#cb39-451"></a>  <span class="fu">group_by</span>(tara_station, name) <span class="sc">%&gt;%</span> </span>
<span id="cb39-452"><a href="#cb39-452"></a>  <span class="fu">count</span>(value) <span class="sc">%&gt;%</span> </span>
<span id="cb39-453"><a href="#cb39-453"></a>  <span class="fu">arrange</span>(name)</span>
<span id="cb39-454"><a href="#cb39-454"></a><span class="in">```</span></span>
<span id="cb39-455"><a href="#cb39-455"></a></span>
<span id="cb39-456"><a href="#cb39-456"></a>Since this is only a few missing observations, we are going to impute them using the <span class="co">[</span><span class="ot">`missRanger` package</span><span class="co">](https://mayer79.github.io/missRanger/index.html)</span>, which uses the <span class="co">[</span><span class="ot">`missForest` random forest imputation approach.</span><span class="co">](https://doi.org/10.1093/bioinformatics/btr597)</span></span>
<span id="cb39-457"><a href="#cb39-457"></a></span>
<span id="cb39-458"><a href="#cb39-458"></a>First only subset variables of interest and make the data unique</span>
<span id="cb39-459"><a href="#cb39-459"></a></span>
<span id="cb39-462"><a href="#cb39-462"></a><span class="in">```{r}</span></span>
<span id="cb39-463"><a href="#cb39-463"></a>tara_biosamples_joined02_replaced02_uniq <span class="ot">&lt;-</span> tara_biosamples_joined02_replaced02 <span class="sc">%&gt;%</span> </span>
<span id="cb39-464"><a href="#cb39-464"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(tara_station<span class="sc">:</span>SiO2) <span class="sc">%&gt;%</span> </span>
<span id="cb39-465"><a href="#cb39-465"></a>  <span class="fu">distinct</span>()</span>
<span id="cb39-466"><a href="#cb39-466"></a><span class="in">```</span></span>
<span id="cb39-467"><a href="#cb39-467"></a></span>
<span id="cb39-468"><a href="#cb39-468"></a>Run the missRanger algorithm to impute missing values. We impute density, temperature, salinity, chl_a by depth, density, temperature, salinity, chl_a, dO2, DIN, NO2, FeT, PO4, SiO2</span>
<span id="cb39-469"><a href="#cb39-469"></a></span>
<span id="cb39-472"><a href="#cb39-472"></a><span class="in">```{r}</span></span>
<span id="cb39-473"><a href="#cb39-473"></a><span class="co">#| output: false</span></span>
<span id="cb39-474"><a href="#cb39-474"></a><span class="co"># set seed for reproducibility</span></span>
<span id="cb39-475"><a href="#cb39-475"></a><span class="fu">set.seed</span>(<span class="dv">35782</span>)</span>
<span id="cb39-476"><a href="#cb39-476"></a>tara_biosamples_joined02_replaced02_uniq_rangerd <span class="ot">&lt;-</span> missRanger<span class="sc">::</span><span class="fu">missRanger</span>(tara_biosamples_joined02_replaced02_uniq, </span>
<span id="cb39-477"><a href="#cb39-477"></a>                                                                           . <span class="sc">~</span> . <span class="sc">-</span> tara_station <span class="sc">-</span> env_ontology <span class="sc">-</span> latitude <span class="sc">-</span> longitude <span class="sc">-</span> date_time <span class="sc">-</span> depth_min <span class="sc">-</span> depth_max, <span class="at">pmm.k =</span> <span class="dv">10</span>, <span class="at">num.trees =</span> <span class="dv">500</span>)</span>
<span id="cb39-478"><a href="#cb39-478"></a><span class="in">```</span></span>
<span id="cb39-479"><a href="#cb39-479"></a></span>
<span id="cb39-480"><a href="#cb39-480"></a><span class="fu">## Joining and replacing missing values with those imputed </span></span>
<span id="cb39-481"><a href="#cb39-481"></a></span>
<span id="cb39-482"><a href="#cb39-482"></a>First get only the samples that needed salinity, temp, density, chl_a replaced from the imputed dataset </span>
<span id="cb39-483"><a href="#cb39-483"></a></span>
<span id="cb39-486"><a href="#cb39-486"></a><span class="in">```{r}</span></span>
<span id="cb39-487"><a href="#cb39-487"></a>tara_biosamples_joined02_replaced02_uniq_imputed <span class="ot">&lt;-</span> tara_biosamples_joined02_replaced02_uniq <span class="sc">%&gt;%</span> <span class="fu">filter</span>(<span class="fu">if_any</span>(<span class="fu">everything</span>(), is.na)) <span class="sc">%&gt;%</span> <span class="fu">select</span>(tara_station<span class="sc">:</span>depth_max)</span>
<span id="cb39-488"><a href="#cb39-488"></a></span>
<span id="cb39-489"><a href="#cb39-489"></a>tara_biosamples_envdata_01 <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(tara_biosamples_joined02_replaced02_uniq_rangerd, tara_biosamples_joined02_replaced02_uniq_imputed,</span>
<span id="cb39-490"><a href="#cb39-490"></a>           <span class="at">by =</span> <span class="fu">join_by</span>(tara_station, env_ontology, latitude, longitude, date_time, depth, depth_min, depth_max))</span>
<span id="cb39-491"><a href="#cb39-491"></a><span class="in">```</span></span>
<span id="cb39-492"><a href="#cb39-492"></a></span>
<span id="cb39-493"><a href="#cb39-493"></a>Next join those imputed samples to the full <span class="in">`tara_biosamples_joined02_replaced02`</span> table</span>
<span id="cb39-494"><a href="#cb39-494"></a></span>
<span id="cb39-497"><a href="#cb39-497"></a><span class="in">```{r}</span></span>
<span id="cb39-498"><a href="#cb39-498"></a>tara_biosamples_envdata_02 <span class="ot">&lt;-</span> <span class="fu">right_join</span>(tara_biosamples_envdata_01, tara_biosamples_joined02_replaced02,</span>
<span id="cb39-499"><a href="#cb39-499"></a>              <span class="at">by =</span> <span class="fu">join_by</span>(tara_station, env_ontology, latitude, longitude, date_time, depth, depth_min, depth_max))</span>
<span id="cb39-500"><a href="#cb39-500"></a><span class="in">```</span></span>
<span id="cb39-501"><a href="#cb39-501"></a></span>
<span id="cb39-502"><a href="#cb39-502"></a><span class="fu">### Detour to add some additional geographic info</span></span>
<span id="cb39-503"><a href="#cb39-503"></a></span>
<span id="cb39-504"><a href="#cb39-504"></a>Next we want to add some geographic information to the dataset like longhurst code, ocean identifier, and the distance to the coast. We will calculate the distance to the nearest coast using <span class="co">[</span><span class="ot">`ggOceanMaps::dist2land`</span><span class="co">](https://mikkovihtakari.github.io/ggOceanMaps/reference/dist2land.html)</span>. <span class="co">[</span><span class="ot">Inspired by this blogpost</span><span class="co">](https://blogs.ubc.ca/yiwang28/2024/01/06/my-r-learning-notes-3-how-to-calculate-the-shortest-distance-to-coast/)</span></span>
<span id="cb39-505"><a href="#cb39-505"></a></span>
<span id="cb39-508"><a href="#cb39-508"></a><span class="in">```{r}</span></span>
<span id="cb39-509"><a href="#cb39-509"></a><span class="co">#| output: false</span></span>
<span id="cb39-510"><a href="#cb39-510"></a><span class="fu">library</span>(ggOceanMaps)</span>
<span id="cb39-511"><a href="#cb39-511"></a><span class="fu">library</span>(rnaturalearth)</span>
<span id="cb39-512"><a href="#cb39-512"></a><span class="fu">library</span>(sf)</span>
<span id="cb39-513"><a href="#cb39-513"></a><span class="in">```</span></span>
<span id="cb39-514"><a href="#cb39-514"></a></span>
<span id="cb39-515"><a href="#cb39-515"></a>Load coastline data at the largest most coarse scale (110) and also subset <span class="in">`PANGAEA_842237`</span> to only coordinates and station identifier</span>
<span id="cb39-516"><a href="#cb39-516"></a></span>
<span id="cb39-519"><a href="#cb39-519"></a><span class="in">```{r}</span></span>
<span id="cb39-520"><a href="#cb39-520"></a>coast <span class="ot">&lt;-</span> rnaturalearth<span class="sc">::</span><span class="fu">ne_coastline</span>(<span class="at">scale=</span><span class="dv">110</span>, <span class="at">returnclass =</span> <span class="st">"sf"</span>)</span>
<span id="cb39-521"><a href="#cb39-521"></a>data <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">select</span>(PANGAEA_842237, tara_station, latitude, longitude, )</span>
<span id="cb39-522"><a href="#cb39-522"></a><span class="in">```</span></span>
<span id="cb39-523"><a href="#cb39-523"></a></span>
<span id="cb39-524"><a href="#cb39-524"></a>Plot to check </span>
<span id="cb39-525"><a href="#cb39-525"></a></span>
<span id="cb39-528"><a href="#cb39-528"></a><span class="in">```{r}</span></span>
<span id="cb39-529"><a href="#cb39-529"></a><span class="fu">plot</span>(coast[<span class="st">'featurecla'</span>])</span>
<span id="cb39-530"><a href="#cb39-530"></a><span class="in">```</span></span>
<span id="cb39-531"><a href="#cb39-531"></a>This function calculates great circle spherical distances (in kilometers) from a coordinate in the ocean to the nearest coastline using the <span class="co">[</span><span class="ot">`st_distance`</span><span class="co">](https://r-spatial.github.io/sf/reference/geos_measures.html)</span> function.</span>
<span id="cb39-532"><a href="#cb39-532"></a></span>
<span id="cb39-535"><a href="#cb39-535"></a><span class="in">```{r}</span></span>
<span id="cb39-536"><a href="#cb39-536"></a>dist2land_df <span class="ot">&lt;-</span> ggOceanMaps<span class="sc">::</span><span class="fu">dist2land</span>(data, <span class="at">shapefile =</span> coast) <span class="sc">%&gt;%</span> </span>
<span id="cb39-537"><a href="#cb39-537"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(tara_station, <span class="at">dist_to_coastline =</span> ldist)</span>
<span id="cb39-538"><a href="#cb39-538"></a><span class="in">```</span></span>
<span id="cb39-539"><a href="#cb39-539"></a></span>
<span id="cb39-540"><a href="#cb39-540"></a>Join the distance to coastline to other general large scale features like longhurst provinces</span>
<span id="cb39-541"><a href="#cb39-541"></a></span>
<span id="cb39-544"><a href="#cb39-544"></a><span class="in">```{r}</span></span>
<span id="cb39-545"><a href="#cb39-545"></a>ocean_feats_df <span class="ot">&lt;-</span> PANGAEA_842237 <span class="sc">%&gt;%</span></span>
<span id="cb39-546"><a href="#cb39-546"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(tara_station, <span class="at">ocean_id =</span> PANGAEA<span class="fl">.842237</span>_016, <span class="at">longhurst_biome =</span> PANGAEA<span class="fl">.842237</span>_015, </span>
<span id="cb39-547"><a href="#cb39-547"></a>                <span class="at">longhurst_code =</span> PANGAEA<span class="fl">.842237</span>_018, <span class="at">bathy_depth =</span> PANGAEA<span class="fl">.842237</span>_012) <span class="sc">%&gt;%</span> </span>
<span id="cb39-548"><a href="#cb39-548"></a>  <span class="fu">left_join</span>(dist2land_df, <span class="at">by =</span> <span class="fu">join_by</span>(tara_station))</span>
<span id="cb39-549"><a href="#cb39-549"></a><span class="in">```</span></span>
<span id="cb39-550"><a href="#cb39-550"></a></span>
<span id="cb39-551"><a href="#cb39-551"></a><span class="fu">### Finalize</span></span>
<span id="cb39-552"><a href="#cb39-552"></a></span>
<span id="cb39-553"><a href="#cb39-553"></a>Now do the final joining to the full imputed data set, replace missing values with the imputed values, and select key variables</span>
<span id="cb39-554"><a href="#cb39-554"></a></span>
<span id="cb39-557"><a href="#cb39-557"></a><span class="in">```{r}</span></span>
<span id="cb39-558"><a href="#cb39-558"></a>tara_biosamples_envdata_final <span class="ot">&lt;-</span> <span class="fu">left_join</span>(tara_biosamples_envdata_02, ocean_feats_df, <span class="at">by =</span> <span class="fu">join_by</span>(tara_station)) <span class="sc">%&gt;%</span> </span>
<span id="cb39-559"><a href="#cb39-559"></a>  <span class="fu">simple_replace</span>(<span class="st">"density"</span>, density.y, density.x) <span class="sc">%&gt;%</span> </span>
<span id="cb39-560"><a href="#cb39-560"></a>  <span class="fu">simple_replace</span>(<span class="st">"salinity"</span>, salinity.y, salinity.x) <span class="sc">%&gt;%</span></span>
<span id="cb39-561"><a href="#cb39-561"></a>  <span class="fu">simple_replace</span>(<span class="st">"temperature"</span>, temperature.y, temperature.x) <span class="sc">%&gt;%</span></span>
<span id="cb39-562"><a href="#cb39-562"></a>  <span class="fu">simple_replace</span>(<span class="st">"chl_a"</span>, chl_a.y, chl_a.x) <span class="sc">%&gt;%</span> </span>
<span id="cb39-563"><a href="#cb39-563"></a>  <span class="fu">select</span>(tara_barcode_num, sra_acc_num<span class="sc">:</span>ena_acc_num, tara_station<span class="sc">:</span>date_time, size_low_thresh, size_high_thresh,</span>
<span id="cb39-564"><a href="#cb39-564"></a>         ocean_id, longhurst_code, longhurst_biome, bathy_depth, dist_to_coastline,</span>
<span id="cb39-565"><a href="#cb39-565"></a>         depth<span class="sc">:</span>depth_max, temperature, salinity, density, chl_a,</span>
<span id="cb39-566"><a href="#cb39-566"></a>         <span class="at">dO2 =</span> dO2.y, <span class="at">DIN =</span> DIN.y, <span class="at">NO2 =</span> NO2.y, <span class="at">FeT =</span> FeT.y, <span class="at">PO4 =</span> PO4.y, <span class="at">SiO2 =</span> SiO2.y)</span>
<span id="cb39-567"><a href="#cb39-567"></a><span class="in">```</span></span>
<span id="cb39-568"><a href="#cb39-568"></a></span>
<span id="cb39-569"><a href="#cb39-569"></a>Final check to see if there is missing entries in this dataframe - good to go</span>
<span id="cb39-570"><a href="#cb39-570"></a></span>
<span id="cb39-573"><a href="#cb39-573"></a><span class="in">```{r}</span></span>
<span id="cb39-574"><a href="#cb39-574"></a>tara_biosamples_envdata_final <span class="sc">%&gt;%</span> </span>
<span id="cb39-575"><a href="#cb39-575"></a>  <span class="fu">filter</span>(<span class="fu">if_any</span>(<span class="fu">everything</span>(), is.na))</span>
<span id="cb39-576"><a href="#cb39-576"></a><span class="in">```</span></span>
<span id="cb39-577"><a href="#cb39-577"></a>Save for later</span>
<span id="cb39-578"><a href="#cb39-578"></a></span>
<span id="cb39-581"><a href="#cb39-581"></a><span class="in">```{r}</span></span>
<span id="cb39-582"><a href="#cb39-582"></a><span class="fu">write_tsv</span>(tara_biosamples_envdata_final, here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"data"</span>, <span class="st">"tara"</span>, <span class="st">"tara_biosamples_envdata_final.tsv"</span>))</span>
<span id="cb39-583"><a href="#cb39-583"></a><span class="in">```</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->




</body></html>